// @cloudflare/vitest-pool-workers/dist/pool/index.mjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// <define:VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES>
var define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES_default;
var init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES = __esm({
  "<define:VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES>"() {
    define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES_default = ["node-internal:async_hooks", "node-internal:buffer", "node-internal:crypto", "node-internal:module", "node-internal:process", "node-internal:util", "node-internal:diagnostics_channel", "node-internal:zlib", "node-internal:url", "node-internal:timers", "node-internal:sqlite", "node-internal:dns", "node:_stream_duplex", "node:_stream_passthrough", "node:_stream_readable", "node:_stream_transform", "node:_stream_writable", "node:_tls_common", "node:_tls_wrap", "node:assert", "node:assert/strict", "node:async_hooks", "node:buffer", "node:constants", "node:crypto", "node:diagnostics_channel", "node:dns", "node:dns/promises", "node:events", "node:fs/promises", "node:module", "node:net", "node:path", "node:path/posix", "node:path/win32", "node:querystring", "node:stream", "node:stream/consumers", "node:stream/promises", "node:stream/web", "node:string_decoder", "node:test", "node:timers", "node:timers/promises", "node:tls", "node:url", "node:util", "node:util/types", "node:zlib", "node-internal:constants", "node-internal:crypto_cipher", "node-internal:crypto_dh", "node-internal:crypto_hash", "node-internal:crypto_hkdf", "node-internal:crypto_keys", "node-internal:crypto_pbkdf2", "node-internal:crypto_random", "node-internal:crypto_scrypt", "node-internal:crypto_sign", "node-internal:crypto_spkac", "node-internal:crypto_util", "node-internal:crypto_x509", "node-internal:debuglog", "node-internal:events", "node-internal:internal_assert", "node-internal:internal_assertionerror", "node-internal:internal_buffer", "node-internal:internal_comparisons", "node-internal:internal_diffs", "node-internal:internal_dns", "node-internal:internal_dns_client", "node-internal:internal_dns_constants", "node-internal:internal_dns_promises", "node-internal:internal_errors", "node-internal:internal_fs", "node-internal:internal_fs_callback", "node-internal:internal_fs_constants", "node-internal:internal_fs_promises", "node-internal:internal_fs_streams", "node-internal:internal_fs_sync", "node-internal:internal_fs_utils", "node-internal:internal_http", "node-internal:internal_http2_constants", "node-internal:internal_http_agent", "node-internal:internal_http_client", "node-internal:internal_http_constants", "node-internal:internal_http_incoming", "node-internal:internal_http_outgoing", "node-internal:internal_http_server", "node-internal:internal_http_util", "node-internal:internal_https_agent", "node-internal:internal_https_server", "node-internal:internal_inspect", "node-internal:internal_module", "node-internal:internal_net", "node-internal:internal_path", "node-internal:internal_process", "node-internal:internal_querystring", "node-internal:internal_readline", "node-internal:internal_readline_promises", "node-internal:internal_stringdecoder", "node-internal:internal_timers", "node-internal:internal_timers_global_override", "node-internal:internal_timers_promises", "node-internal:internal_tls", "node-internal:internal_tls_common", "node-internal:internal_tls_constants", "node-internal:internal_tls_jsstream", "node-internal:internal_tls_wrap", "node-internal:internal_types", "node-internal:internal_url", "node-internal:internal_utils", "node-internal:internal_zlib", "node-internal:internal_zlib_base", "node-internal:internal_zlib_constants", "node-internal:legacy_process", "node-internal:legacy_url", "node-internal:mock", "node-internal:public_process", "node-internal:streams_add_abort_signal", "node-internal:streams_compose", "node-internal:streams_destroy", "node-internal:streams_duplex", "node-internal:streams_end_of_stream", "node-internal:streams_legacy", "node-internal:streams_pipeline", "node-internal:streams_promises", "node-internal:streams_readable", "node-internal:streams_state", "node-internal:streams_transform", "node-internal:streams_util", "node-internal:streams_writable", "node-internal:validators", "internal:unsafe-eval", "cloudflare-internal:sockets", "cloudflare-internal:base64", "cloudflare:ai", "cloudflare:br", "cloudflare:email", "cloudflare:node", "cloudflare:pipelines", "cloudflare:sockets", "cloudflare:vectorize", "cloudflare:workers", "cloudflare:workflows", "cloudflare-internal:ai-api", "cloudflare-internal:aig-api", "cloudflare-internal:autorag-api", "cloudflare-internal:br-api", "cloudflare-internal:d1-api", "cloudflare-internal:http", "cloudflare-internal:images-api", "cloudflare-internal:pipeline-transform", "cloudflare-internal:streaming-base64", "cloudflare-internal:streaming-forms", "cloudflare-internal:test-tracing-wrapper", "cloudflare-internal:to-markdown-api", "cloudflare-internal:tracing-helpers", "cloudflare-internal:vectorize-api", "cloudflare-internal:workflows-api", "cloudflare-internal:workers", "cloudflare-internal:tracing", "cloudflare-internal:env", "cloudflare-internal:filesystem", "workerd:unsafe"];
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/lrucache.js
var require_lrucache = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/lrucache.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var LRUCache = class {
      constructor() {
        this.max = 1e3;
        this.map = /* @__PURE__ */ new Map();
      }
      get(key) {
        const value = this.map.get(key);
        if (value === void 0) {
          return void 0;
        } else {
          this.map.delete(key);
          this.map.set(key, value);
          return value;
        }
      }
      delete(key) {
        return this.map.delete(key);
      }
      set(key, value) {
        const deleted = this.delete(key);
        if (!deleted && value !== void 0) {
          if (this.map.size >= this.max) {
            const firstKey = this.map.keys().next().value;
            this.delete(firstKey);
          }
          this.map.set(key, value);
        }
        return this;
      }
    };
    module.exports = LRUCache;
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/parse-options.js
var require_parse_options = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/parse-options.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var looseOption = Object.freeze({ loose: true });
    var emptyOpts = Object.freeze({});
    var parseOptions = (options) => {
      if (!options) {
        return emptyOpts;
      }
      if (typeof options !== "object") {
        return looseOption;
      }
      return options;
    };
    module.exports = parseOptions;
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/constants.js
var require_constants = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/constants.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var SEMVER_SPEC_VERSION = "2.0.0";
    var MAX_LENGTH = 256;
    var MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || /* istanbul ignore next */
    9007199254740991;
    var MAX_SAFE_COMPONENT_LENGTH = 16;
    var MAX_SAFE_BUILD_LENGTH = MAX_LENGTH - 6;
    var RELEASE_TYPES = [
      "major",
      "premajor",
      "minor",
      "preminor",
      "patch",
      "prepatch",
      "prerelease"
    ];
    module.exports = {
      MAX_LENGTH,
      MAX_SAFE_COMPONENT_LENGTH,
      MAX_SAFE_BUILD_LENGTH,
      MAX_SAFE_INTEGER,
      RELEASE_TYPES,
      SEMVER_SPEC_VERSION,
      FLAG_INCLUDE_PRERELEASE: 1,
      FLAG_LOOSE: 2
    };
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/debug.js
var require_debug = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/debug.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var debug = typeof process === "object" && process.env && process.env.NODE_DEBUG && /\bsemver\b/i.test(process.env.NODE_DEBUG) ? (...args) => console.error("SEMVER", ...args) : () => {
    };
    module.exports = debug;
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/re.js
var require_re = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/re.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var {
      MAX_SAFE_COMPONENT_LENGTH,
      MAX_SAFE_BUILD_LENGTH,
      MAX_LENGTH
    } = require_constants();
    var debug = require_debug();
    exports = module.exports = {};
    var re = exports.re = [];
    var safeRe = exports.safeRe = [];
    var src = exports.src = [];
    var safeSrc = exports.safeSrc = [];
    var t = exports.t = {};
    var R = 0;
    var LETTERDASHNUMBER = "[a-zA-Z0-9-]";
    var safeRegexReplacements = [
      ["\\s", 1],
      ["\\d", MAX_LENGTH],
      [LETTERDASHNUMBER, MAX_SAFE_BUILD_LENGTH]
    ];
    var makeSafeRegex = (value) => {
      for (const [token, max] of safeRegexReplacements) {
        value = value.split(`${token}*`).join(`${token}{0,${max}}`).split(`${token}+`).join(`${token}{1,${max}}`);
      }
      return value;
    };
    var createToken = (name, value, isGlobal) => {
      const safe = makeSafeRegex(value);
      const index = R++;
      debug(name, index, value);
      t[name] = index;
      src[index] = value;
      safeSrc[index] = safe;
      re[index] = new RegExp(value, isGlobal ? "g" : void 0);
      safeRe[index] = new RegExp(safe, isGlobal ? "g" : void 0);
    };
    createToken("NUMERICIDENTIFIER", "0|[1-9]\\d*");
    createToken("NUMERICIDENTIFIERLOOSE", "\\d+");
    createToken("NONNUMERICIDENTIFIER", `\\d*[a-zA-Z-]${LETTERDASHNUMBER}*`);
    createToken("MAINVERSION", `(${src[t.NUMERICIDENTIFIER]})\\.(${src[t.NUMERICIDENTIFIER]})\\.(${src[t.NUMERICIDENTIFIER]})`);
    createToken("MAINVERSIONLOOSE", `(${src[t.NUMERICIDENTIFIERLOOSE]})\\.(${src[t.NUMERICIDENTIFIERLOOSE]})\\.(${src[t.NUMERICIDENTIFIERLOOSE]})`);
    createToken("PRERELEASEIDENTIFIER", `(?:${src[t.NUMERICIDENTIFIER]}|${src[t.NONNUMERICIDENTIFIER]})`);
    createToken("PRERELEASEIDENTIFIERLOOSE", `(?:${src[t.NUMERICIDENTIFIERLOOSE]}|${src[t.NONNUMERICIDENTIFIER]})`);
    createToken("PRERELEASE", `(?:-(${src[t.PRERELEASEIDENTIFIER]}(?:\\.${src[t.PRERELEASEIDENTIFIER]})*))`);
    createToken("PRERELEASELOOSE", `(?:-?(${src[t.PRERELEASEIDENTIFIERLOOSE]}(?:\\.${src[t.PRERELEASEIDENTIFIERLOOSE]})*))`);
    createToken("BUILDIDENTIFIER", `${LETTERDASHNUMBER}+`);
    createToken("BUILD", `(?:\\+(${src[t.BUILDIDENTIFIER]}(?:\\.${src[t.BUILDIDENTIFIER]})*))`);
    createToken("FULLPLAIN", `v?${src[t.MAINVERSION]}${src[t.PRERELEASE]}?${src[t.BUILD]}?`);
    createToken("FULL", `^${src[t.FULLPLAIN]}$`);
    createToken("LOOSEPLAIN", `[v=\\s]*${src[t.MAINVERSIONLOOSE]}${src[t.PRERELEASELOOSE]}?${src[t.BUILD]}?`);
    createToken("LOOSE", `^${src[t.LOOSEPLAIN]}$`);
    createToken("GTLT", "((?:<|>)?=?)");
    createToken("XRANGEIDENTIFIERLOOSE", `${src[t.NUMERICIDENTIFIERLOOSE]}|x|X|\\*`);
    createToken("XRANGEIDENTIFIER", `${src[t.NUMERICIDENTIFIER]}|x|X|\\*`);
    createToken("XRANGEPLAIN", `[v=\\s]*(${src[t.XRANGEIDENTIFIER]})(?:\\.(${src[t.XRANGEIDENTIFIER]})(?:\\.(${src[t.XRANGEIDENTIFIER]})(?:${src[t.PRERELEASE]})?${src[t.BUILD]}?)?)?`);
    createToken("XRANGEPLAINLOOSE", `[v=\\s]*(${src[t.XRANGEIDENTIFIERLOOSE]})(?:\\.(${src[t.XRANGEIDENTIFIERLOOSE]})(?:\\.(${src[t.XRANGEIDENTIFIERLOOSE]})(?:${src[t.PRERELEASELOOSE]})?${src[t.BUILD]}?)?)?`);
    createToken("XRANGE", `^${src[t.GTLT]}\\s*${src[t.XRANGEPLAIN]}$`);
    createToken("XRANGELOOSE", `^${src[t.GTLT]}\\s*${src[t.XRANGEPLAINLOOSE]}$`);
    createToken("COERCEPLAIN", `${"(^|[^\\d])(\\d{1,"}${MAX_SAFE_COMPONENT_LENGTH}})(?:\\.(\\d{1,${MAX_SAFE_COMPONENT_LENGTH}}))?(?:\\.(\\d{1,${MAX_SAFE_COMPONENT_LENGTH}}))?`);
    createToken("COERCE", `${src[t.COERCEPLAIN]}(?:$|[^\\d])`);
    createToken("COERCEFULL", src[t.COERCEPLAIN] + `(?:${src[t.PRERELEASE]})?(?:${src[t.BUILD]})?(?:$|[^\\d])`);
    createToken("COERCERTL", src[t.COERCE], true);
    createToken("COERCERTLFULL", src[t.COERCEFULL], true);
    createToken("LONETILDE", "(?:~>?)");
    createToken("TILDETRIM", `(\\s*)${src[t.LONETILDE]}\\s+`, true);
    exports.tildeTrimReplace = "$1~";
    createToken("TILDE", `^${src[t.LONETILDE]}${src[t.XRANGEPLAIN]}$`);
    createToken("TILDELOOSE", `^${src[t.LONETILDE]}${src[t.XRANGEPLAINLOOSE]}$`);
    createToken("LONECARET", "(?:\\^)");
    createToken("CARETTRIM", `(\\s*)${src[t.LONECARET]}\\s+`, true);
    exports.caretTrimReplace = "$1^";
    createToken("CARET", `^${src[t.LONECARET]}${src[t.XRANGEPLAIN]}$`);
    createToken("CARETLOOSE", `^${src[t.LONECARET]}${src[t.XRANGEPLAINLOOSE]}$`);
    createToken("COMPARATORLOOSE", `^${src[t.GTLT]}\\s*(${src[t.LOOSEPLAIN]})$|^$`);
    createToken("COMPARATOR", `^${src[t.GTLT]}\\s*(${src[t.FULLPLAIN]})$|^$`);
    createToken("COMPARATORTRIM", `(\\s*)${src[t.GTLT]}\\s*(${src[t.LOOSEPLAIN]}|${src[t.XRANGEPLAIN]})`, true);
    exports.comparatorTrimReplace = "$1$2$3";
    createToken("HYPHENRANGE", `^\\s*(${src[t.XRANGEPLAIN]})\\s+-\\s+(${src[t.XRANGEPLAIN]})\\s*$`);
    createToken("HYPHENRANGELOOSE", `^\\s*(${src[t.XRANGEPLAINLOOSE]})\\s+-\\s+(${src[t.XRANGEPLAINLOOSE]})\\s*$`);
    createToken("STAR", "(<|>)?=?\\s*\\*");
    createToken("GTE0", "^\\s*>=\\s*0\\.0\\.0\\s*$");
    createToken("GTE0PRE", "^\\s*>=\\s*0\\.0\\.0-0\\s*$");
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/identifiers.js
var require_identifiers = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/identifiers.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var numeric = /^[0-9]+$/;
    var compareIdentifiers = (a, b) => {
      const anum = numeric.test(a);
      const bnum = numeric.test(b);
      if (anum && bnum) {
        a = +a;
        b = +b;
      }
      return a === b ? 0 : anum && !bnum ? -1 : bnum && !anum ? 1 : a < b ? -1 : 1;
    };
    var rcompareIdentifiers = (a, b) => compareIdentifiers(b, a);
    module.exports = {
      compareIdentifiers,
      rcompareIdentifiers
    };
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/classes/semver.js
var require_semver = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/classes/semver.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var debug = require_debug();
    var { MAX_LENGTH, MAX_SAFE_INTEGER } = require_constants();
    var { safeRe: re, safeSrc: src, t } = require_re();
    var parseOptions = require_parse_options();
    var { compareIdentifiers } = require_identifiers();
    var SemVer = class _SemVer {
      constructor(version, options) {
        options = parseOptions(options);
        if (version instanceof _SemVer) {
          if (version.loose === !!options.loose && version.includePrerelease === !!options.includePrerelease) {
            return version;
          } else {
            version = version.version;
          }
        } else if (typeof version !== "string") {
          throw new TypeError(`Invalid version. Must be a string. Got type "${typeof version}".`);
        }
        if (version.length > MAX_LENGTH) {
          throw new TypeError(
            `version is longer than ${MAX_LENGTH} characters`
          );
        }
        debug("SemVer", version, options);
        this.options = options;
        this.loose = !!options.loose;
        this.includePrerelease = !!options.includePrerelease;
        const m = version.trim().match(options.loose ? re[t.LOOSE] : re[t.FULL]);
        if (!m) {
          throw new TypeError(`Invalid Version: ${version}`);
        }
        this.raw = version;
        this.major = +m[1];
        this.minor = +m[2];
        this.patch = +m[3];
        if (this.major > MAX_SAFE_INTEGER || this.major < 0) {
          throw new TypeError("Invalid major version");
        }
        if (this.minor > MAX_SAFE_INTEGER || this.minor < 0) {
          throw new TypeError("Invalid minor version");
        }
        if (this.patch > MAX_SAFE_INTEGER || this.patch < 0) {
          throw new TypeError("Invalid patch version");
        }
        if (!m[4]) {
          this.prerelease = [];
        } else {
          this.prerelease = m[4].split(".").map((id) => {
            if (/^[0-9]+$/.test(id)) {
              const num = +id;
              if (num >= 0 && num < MAX_SAFE_INTEGER) {
                return num;
              }
            }
            return id;
          });
        }
        this.build = m[5] ? m[5].split(".") : [];
        this.format();
      }
      format() {
        this.version = `${this.major}.${this.minor}.${this.patch}`;
        if (this.prerelease.length) {
          this.version += `-${this.prerelease.join(".")}`;
        }
        return this.version;
      }
      toString() {
        return this.version;
      }
      compare(other) {
        debug("SemVer.compare", this.version, this.options, other);
        if (!(other instanceof _SemVer)) {
          if (typeof other === "string" && other === this.version) {
            return 0;
          }
          other = new _SemVer(other, this.options);
        }
        if (other.version === this.version) {
          return 0;
        }
        return this.compareMain(other) || this.comparePre(other);
      }
      compareMain(other) {
        if (!(other instanceof _SemVer)) {
          other = new _SemVer(other, this.options);
        }
        return compareIdentifiers(this.major, other.major) || compareIdentifiers(this.minor, other.minor) || compareIdentifiers(this.patch, other.patch);
      }
      comparePre(other) {
        if (!(other instanceof _SemVer)) {
          other = new _SemVer(other, this.options);
        }
        if (this.prerelease.length && !other.prerelease.length) {
          return -1;
        } else if (!this.prerelease.length && other.prerelease.length) {
          return 1;
        } else if (!this.prerelease.length && !other.prerelease.length) {
          return 0;
        }
        let i = 0;
        do {
          const a = this.prerelease[i];
          const b = other.prerelease[i];
          debug("prerelease compare", i, a, b);
          if (a === void 0 && b === void 0) {
            return 0;
          } else if (b === void 0) {
            return 1;
          } else if (a === void 0) {
            return -1;
          } else if (a === b) {
            continue;
          } else {
            return compareIdentifiers(a, b);
          }
        } while (++i);
      }
      compareBuild(other) {
        if (!(other instanceof _SemVer)) {
          other = new _SemVer(other, this.options);
        }
        let i = 0;
        do {
          const a = this.build[i];
          const b = other.build[i];
          debug("build compare", i, a, b);
          if (a === void 0 && b === void 0) {
            return 0;
          } else if (b === void 0) {
            return 1;
          } else if (a === void 0) {
            return -1;
          } else if (a === b) {
            continue;
          } else {
            return compareIdentifiers(a, b);
          }
        } while (++i);
      }
      // preminor will bump the version up to the next minor release, and immediately
      // down to pre-release. premajor and prepatch work the same way.
      inc(release, identifier, identifierBase) {
        if (release.startsWith("pre")) {
          if (!identifier && identifierBase === false) {
            throw new Error("invalid increment argument: identifier is empty");
          }
          if (identifier) {
            const r = new RegExp(`^${this.options.loose ? src[t.PRERELEASELOOSE] : src[t.PRERELEASE]}$`);
            const match = `-${identifier}`.match(r);
            if (!match || match[1] !== identifier) {
              throw new Error(`invalid identifier: ${identifier}`);
            }
          }
        }
        switch (release) {
          case "premajor":
            this.prerelease.length = 0;
            this.patch = 0;
            this.minor = 0;
            this.major++;
            this.inc("pre", identifier, identifierBase);
            break;
          case "preminor":
            this.prerelease.length = 0;
            this.patch = 0;
            this.minor++;
            this.inc("pre", identifier, identifierBase);
            break;
          case "prepatch":
            this.prerelease.length = 0;
            this.inc("patch", identifier, identifierBase);
            this.inc("pre", identifier, identifierBase);
            break;
          // If the input is a non-prerelease version, this acts the same as
          // prepatch.
          case "prerelease":
            if (this.prerelease.length === 0) {
              this.inc("patch", identifier, identifierBase);
            }
            this.inc("pre", identifier, identifierBase);
            break;
          case "release":
            if (this.prerelease.length === 0) {
              throw new Error(`version ${this.raw} is not a prerelease`);
            }
            this.prerelease.length = 0;
            break;
          case "major":
            if (this.minor !== 0 || this.patch !== 0 || this.prerelease.length === 0) {
              this.major++;
            }
            this.minor = 0;
            this.patch = 0;
            this.prerelease = [];
            break;
          case "minor":
            if (this.patch !== 0 || this.prerelease.length === 0) {
              this.minor++;
            }
            this.patch = 0;
            this.prerelease = [];
            break;
          case "patch":
            if (this.prerelease.length === 0) {
              this.patch++;
            }
            this.prerelease = [];
            break;
          // This probably shouldn't be used publicly.
          // 1.0.0 'pre' would become 1.0.0-0 which is the wrong direction.
          case "pre": {
            const base = Number(identifierBase) ? 1 : 0;
            if (this.prerelease.length === 0) {
              this.prerelease = [base];
            } else {
              let i = this.prerelease.length;
              while (--i >= 0) {
                if (typeof this.prerelease[i] === "number") {
                  this.prerelease[i]++;
                  i = -2;
                }
              }
              if (i === -1) {
                if (identifier === this.prerelease.join(".") && identifierBase === false) {
                  throw new Error("invalid increment argument: identifier already exists");
                }
                this.prerelease.push(base);
              }
            }
            if (identifier) {
              let prerelease = [identifier, base];
              if (identifierBase === false) {
                prerelease = [identifier];
              }
              if (compareIdentifiers(this.prerelease[0], identifier) === 0) {
                if (isNaN(this.prerelease[1])) {
                  this.prerelease = prerelease;
                }
              } else {
                this.prerelease = prerelease;
              }
            }
            break;
          }
          default:
            throw new Error(`invalid increment argument: ${release}`);
        }
        this.raw = this.format();
        if (this.build.length) {
          this.raw += `+${this.build.join(".")}`;
        }
        return this;
      }
    };
    module.exports = SemVer;
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/compare.js
var require_compare = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/compare.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var SemVer = require_semver();
    var compare = (a, b, loose) => new SemVer(a, loose).compare(new SemVer(b, loose));
    module.exports = compare;
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/eq.js
var require_eq = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/eq.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var compare = require_compare();
    var eq = (a, b, loose) => compare(a, b, loose) === 0;
    module.exports = eq;
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/neq.js
var require_neq = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/neq.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var compare = require_compare();
    var neq = (a, b, loose) => compare(a, b, loose) !== 0;
    module.exports = neq;
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/gt.js
var require_gt = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/gt.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var compare = require_compare();
    var gt = (a, b, loose) => compare(a, b, loose) > 0;
    module.exports = gt;
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/gte.js
var require_gte = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/gte.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var compare = require_compare();
    var gte = (a, b, loose) => compare(a, b, loose) >= 0;
    module.exports = gte;
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/lt.js
var require_lt = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/lt.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var compare = require_compare();
    var lt = (a, b, loose) => compare(a, b, loose) < 0;
    module.exports = lt;
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/lte.js
var require_lte = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/lte.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var compare = require_compare();
    var lte = (a, b, loose) => compare(a, b, loose) <= 0;
    module.exports = lte;
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/cmp.js
var require_cmp = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/cmp.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var eq = require_eq();
    var neq = require_neq();
    var gt = require_gt();
    var gte = require_gte();
    var lt = require_lt();
    var lte = require_lte();
    var cmp = (a, op, b, loose) => {
      switch (op) {
        case "===":
          if (typeof a === "object") {
            a = a.version;
          }
          if (typeof b === "object") {
            b = b.version;
          }
          return a === b;
        case "!==":
          if (typeof a === "object") {
            a = a.version;
          }
          if (typeof b === "object") {
            b = b.version;
          }
          return a !== b;
        case "":
        case "=":
        case "==":
          return eq(a, b, loose);
        case "!=":
          return neq(a, b, loose);
        case ">":
          return gt(a, b, loose);
        case ">=":
          return gte(a, b, loose);
        case "<":
          return lt(a, b, loose);
        case "<=":
          return lte(a, b, loose);
        default:
          throw new TypeError(`Invalid operator: ${op}`);
      }
    };
    module.exports = cmp;
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/classes/comparator.js
var require_comparator = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/classes/comparator.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var ANY = Symbol("SemVer ANY");
    var Comparator = class _Comparator {
      static get ANY() {
        return ANY;
      }
      constructor(comp, options) {
        options = parseOptions(options);
        if (comp instanceof _Comparator) {
          if (comp.loose === !!options.loose) {
            return comp;
          } else {
            comp = comp.value;
          }
        }
        comp = comp.trim().split(/\s+/).join(" ");
        debug("comparator", comp, options);
        this.options = options;
        this.loose = !!options.loose;
        this.parse(comp);
        if (this.semver === ANY) {
          this.value = "";
        } else {
          this.value = this.operator + this.semver.version;
        }
        debug("comp", this);
      }
      parse(comp) {
        const r = this.options.loose ? re[t.COMPARATORLOOSE] : re[t.COMPARATOR];
        const m = comp.match(r);
        if (!m) {
          throw new TypeError(`Invalid comparator: ${comp}`);
        }
        this.operator = m[1] !== void 0 ? m[1] : "";
        if (this.operator === "=") {
          this.operator = "";
        }
        if (!m[2]) {
          this.semver = ANY;
        } else {
          this.semver = new SemVer(m[2], this.options.loose);
        }
      }
      toString() {
        return this.value;
      }
      test(version) {
        debug("Comparator.test", version, this.options.loose);
        if (this.semver === ANY || version === ANY) {
          return true;
        }
        if (typeof version === "string") {
          try {
            version = new SemVer(version, this.options);
          } catch (er) {
            return false;
          }
        }
        return cmp(version, this.operator, this.semver, this.options);
      }
      intersects(comp, options) {
        if (!(comp instanceof _Comparator)) {
          throw new TypeError("a Comparator is required");
        }
        if (this.operator === "") {
          if (this.value === "") {
            return true;
          }
          return new Range(comp.value, options).test(this.value);
        } else if (comp.operator === "") {
          if (comp.value === "") {
            return true;
          }
          return new Range(this.value, options).test(comp.semver);
        }
        options = parseOptions(options);
        if (options.includePrerelease && (this.value === "<0.0.0-0" || comp.value === "<0.0.0-0")) {
          return false;
        }
        if (!options.includePrerelease && (this.value.startsWith("<0.0.0") || comp.value.startsWith("<0.0.0"))) {
          return false;
        }
        if (this.operator.startsWith(">") && comp.operator.startsWith(">")) {
          return true;
        }
        if (this.operator.startsWith("<") && comp.operator.startsWith("<")) {
          return true;
        }
        if (this.semver.version === comp.semver.version && this.operator.includes("=") && comp.operator.includes("=")) {
          return true;
        }
        if (cmp(this.semver, "<", comp.semver, options) && this.operator.startsWith(">") && comp.operator.startsWith("<")) {
          return true;
        }
        if (cmp(this.semver, ">", comp.semver, options) && this.operator.startsWith("<") && comp.operator.startsWith(">")) {
          return true;
        }
        return false;
      }
    };
    module.exports = Comparator;
    var parseOptions = require_parse_options();
    var { safeRe: re, t } = require_re();
    var cmp = require_cmp();
    var debug = require_debug();
    var SemVer = require_semver();
    var Range = require_range();
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/classes/range.js
var require_range = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/classes/range.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var SPACE_CHARACTERS = /\s+/g;
    var Range = class _Range {
      constructor(range, options) {
        options = parseOptions(options);
        if (range instanceof _Range) {
          if (range.loose === !!options.loose && range.includePrerelease === !!options.includePrerelease) {
            return range;
          } else {
            return new _Range(range.raw, options);
          }
        }
        if (range instanceof Comparator) {
          this.raw = range.value;
          this.set = [[range]];
          this.formatted = void 0;
          return this;
        }
        this.options = options;
        this.loose = !!options.loose;
        this.includePrerelease = !!options.includePrerelease;
        this.raw = range.trim().replace(SPACE_CHARACTERS, " ");
        this.set = this.raw.split("||").map((r) => this.parseRange(r.trim())).filter((c) => c.length);
        if (!this.set.length) {
          throw new TypeError(`Invalid SemVer Range: ${this.raw}`);
        }
        if (this.set.length > 1) {
          const first = this.set[0];
          this.set = this.set.filter((c) => !isNullSet(c[0]));
          if (this.set.length === 0) {
            this.set = [first];
          } else if (this.set.length > 1) {
            for (const c of this.set) {
              if (c.length === 1 && isAny(c[0])) {
                this.set = [c];
                break;
              }
            }
          }
        }
        this.formatted = void 0;
      }
      get range() {
        if (this.formatted === void 0) {
          this.formatted = "";
          for (let i = 0; i < this.set.length; i++) {
            if (i > 0) {
              this.formatted += "||";
            }
            const comps = this.set[i];
            for (let k = 0; k < comps.length; k++) {
              if (k > 0) {
                this.formatted += " ";
              }
              this.formatted += comps[k].toString().trim();
            }
          }
        }
        return this.formatted;
      }
      format() {
        return this.range;
      }
      toString() {
        return this.range;
      }
      parseRange(range) {
        const memoOpts = (this.options.includePrerelease && FLAG_INCLUDE_PRERELEASE) | (this.options.loose && FLAG_LOOSE);
        const memoKey = memoOpts + ":" + range;
        const cached = cache.get(memoKey);
        if (cached) {
          return cached;
        }
        const loose = this.options.loose;
        const hr = loose ? re[t.HYPHENRANGELOOSE] : re[t.HYPHENRANGE];
        range = range.replace(hr, hyphenReplace(this.options.includePrerelease));
        debug("hyphen replace", range);
        range = range.replace(re[t.COMPARATORTRIM], comparatorTrimReplace);
        debug("comparator trim", range);
        range = range.replace(re[t.TILDETRIM], tildeTrimReplace);
        debug("tilde trim", range);
        range = range.replace(re[t.CARETTRIM], caretTrimReplace);
        debug("caret trim", range);
        let rangeList = range.split(" ").map((comp) => parseComparator(comp, this.options)).join(" ").split(/\s+/).map((comp) => replaceGTE0(comp, this.options));
        if (loose) {
          rangeList = rangeList.filter((comp) => {
            debug("loose invalid filter", comp, this.options);
            return !!comp.match(re[t.COMPARATORLOOSE]);
          });
        }
        debug("range list", rangeList);
        const rangeMap = /* @__PURE__ */ new Map();
        const comparators = rangeList.map((comp) => new Comparator(comp, this.options));
        for (const comp of comparators) {
          if (isNullSet(comp)) {
            return [comp];
          }
          rangeMap.set(comp.value, comp);
        }
        if (rangeMap.size > 1 && rangeMap.has("")) {
          rangeMap.delete("");
        }
        const result = [...rangeMap.values()];
        cache.set(memoKey, result);
        return result;
      }
      intersects(range, options) {
        if (!(range instanceof _Range)) {
          throw new TypeError("a Range is required");
        }
        return this.set.some((thisComparators) => {
          return isSatisfiable(thisComparators, options) && range.set.some((rangeComparators) => {
            return isSatisfiable(rangeComparators, options) && thisComparators.every((thisComparator) => {
              return rangeComparators.every((rangeComparator) => {
                return thisComparator.intersects(rangeComparator, options);
              });
            });
          });
        });
      }
      // if ANY of the sets match ALL of its comparators, then pass
      test(version) {
        if (!version) {
          return false;
        }
        if (typeof version === "string") {
          try {
            version = new SemVer(version, this.options);
          } catch (er) {
            return false;
          }
        }
        for (let i = 0; i < this.set.length; i++) {
          if (testSet(this.set[i], version, this.options)) {
            return true;
          }
        }
        return false;
      }
    };
    module.exports = Range;
    var LRU = require_lrucache();
    var cache = new LRU();
    var parseOptions = require_parse_options();
    var Comparator = require_comparator();
    var debug = require_debug();
    var SemVer = require_semver();
    var {
      safeRe: re,
      t,
      comparatorTrimReplace,
      tildeTrimReplace,
      caretTrimReplace
    } = require_re();
    var { FLAG_INCLUDE_PRERELEASE, FLAG_LOOSE } = require_constants();
    var isNullSet = (c) => c.value === "<0.0.0-0";
    var isAny = (c) => c.value === "";
    var isSatisfiable = (comparators, options) => {
      let result = true;
      const remainingComparators = comparators.slice();
      let testComparator = remainingComparators.pop();
      while (result && remainingComparators.length) {
        result = remainingComparators.every((otherComparator) => {
          return testComparator.intersects(otherComparator, options);
        });
        testComparator = remainingComparators.pop();
      }
      return result;
    };
    var parseComparator = (comp, options) => {
      debug("comp", comp, options);
      comp = replaceCarets(comp, options);
      debug("caret", comp);
      comp = replaceTildes(comp, options);
      debug("tildes", comp);
      comp = replaceXRanges(comp, options);
      debug("xrange", comp);
      comp = replaceStars(comp, options);
      debug("stars", comp);
      return comp;
    };
    var isX = (id) => !id || id.toLowerCase() === "x" || id === "*";
    var replaceTildes = (comp, options) => {
      return comp.trim().split(/\s+/).map((c) => replaceTilde(c, options)).join(" ");
    };
    var replaceTilde = (comp, options) => {
      const r = options.loose ? re[t.TILDELOOSE] : re[t.TILDE];
      return comp.replace(r, (_, M, m, p, pr) => {
        debug("tilde", comp, _, M, m, p, pr);
        let ret;
        if (isX(M)) {
          ret = "";
        } else if (isX(m)) {
          ret = `>=${M}.0.0 <${+M + 1}.0.0-0`;
        } else if (isX(p)) {
          ret = `>=${M}.${m}.0 <${M}.${+m + 1}.0-0`;
        } else if (pr) {
          debug("replaceTilde pr", pr);
          ret = `>=${M}.${m}.${p}-${pr} <${M}.${+m + 1}.0-0`;
        } else {
          ret = `>=${M}.${m}.${p} <${M}.${+m + 1}.0-0`;
        }
        debug("tilde return", ret);
        return ret;
      });
    };
    var replaceCarets = (comp, options) => {
      return comp.trim().split(/\s+/).map((c) => replaceCaret(c, options)).join(" ");
    };
    var replaceCaret = (comp, options) => {
      debug("caret", comp, options);
      const r = options.loose ? re[t.CARETLOOSE] : re[t.CARET];
      const z2 = options.includePrerelease ? "-0" : "";
      return comp.replace(r, (_, M, m, p, pr) => {
        debug("caret", comp, _, M, m, p, pr);
        let ret;
        if (isX(M)) {
          ret = "";
        } else if (isX(m)) {
          ret = `>=${M}.0.0${z2} <${+M + 1}.0.0-0`;
        } else if (isX(p)) {
          if (M === "0") {
            ret = `>=${M}.${m}.0${z2} <${M}.${+m + 1}.0-0`;
          } else {
            ret = `>=${M}.${m}.0${z2} <${+M + 1}.0.0-0`;
          }
        } else if (pr) {
          debug("replaceCaret pr", pr);
          if (M === "0") {
            if (m === "0") {
              ret = `>=${M}.${m}.${p}-${pr} <${M}.${m}.${+p + 1}-0`;
            } else {
              ret = `>=${M}.${m}.${p}-${pr} <${M}.${+m + 1}.0-0`;
            }
          } else {
            ret = `>=${M}.${m}.${p}-${pr} <${+M + 1}.0.0-0`;
          }
        } else {
          debug("no pr");
          if (M === "0") {
            if (m === "0") {
              ret = `>=${M}.${m}.${p}${z2} <${M}.${m}.${+p + 1}-0`;
            } else {
              ret = `>=${M}.${m}.${p}${z2} <${M}.${+m + 1}.0-0`;
            }
          } else {
            ret = `>=${M}.${m}.${p} <${+M + 1}.0.0-0`;
          }
        }
        debug("caret return", ret);
        return ret;
      });
    };
    var replaceXRanges = (comp, options) => {
      debug("replaceXRanges", comp, options);
      return comp.split(/\s+/).map((c) => replaceXRange(c, options)).join(" ");
    };
    var replaceXRange = (comp, options) => {
      comp = comp.trim();
      const r = options.loose ? re[t.XRANGELOOSE] : re[t.XRANGE];
      return comp.replace(r, (ret, gtlt, M, m, p, pr) => {
        debug("xRange", comp, ret, gtlt, M, m, p, pr);
        const xM = isX(M);
        const xm = xM || isX(m);
        const xp = xm || isX(p);
        const anyX = xp;
        if (gtlt === "=" && anyX) {
          gtlt = "";
        }
        pr = options.includePrerelease ? "-0" : "";
        if (xM) {
          if (gtlt === ">" || gtlt === "<") {
            ret = "<0.0.0-0";
          } else {
            ret = "*";
          }
        } else if (gtlt && anyX) {
          if (xm) {
            m = 0;
          }
          p = 0;
          if (gtlt === ">") {
            gtlt = ">=";
            if (xm) {
              M = +M + 1;
              m = 0;
              p = 0;
            } else {
              m = +m + 1;
              p = 0;
            }
          } else if (gtlt === "<=") {
            gtlt = "<";
            if (xm) {
              M = +M + 1;
            } else {
              m = +m + 1;
            }
          }
          if (gtlt === "<") {
            pr = "-0";
          }
          ret = `${gtlt + M}.${m}.${p}${pr}`;
        } else if (xm) {
          ret = `>=${M}.0.0${pr} <${+M + 1}.0.0-0`;
        } else if (xp) {
          ret = `>=${M}.${m}.0${pr} <${M}.${+m + 1}.0-0`;
        }
        debug("xRange return", ret);
        return ret;
      });
    };
    var replaceStars = (comp, options) => {
      debug("replaceStars", comp, options);
      return comp.trim().replace(re[t.STAR], "");
    };
    var replaceGTE0 = (comp, options) => {
      debug("replaceGTE0", comp, options);
      return comp.trim().replace(re[options.includePrerelease ? t.GTE0PRE : t.GTE0], "");
    };
    var hyphenReplace = (incPr) => ($0, from, fM, fm, fp, fpr, fb, to, tM, tm, tp, tpr) => {
      if (isX(fM)) {
        from = "";
      } else if (isX(fm)) {
        from = `>=${fM}.0.0${incPr ? "-0" : ""}`;
      } else if (isX(fp)) {
        from = `>=${fM}.${fm}.0${incPr ? "-0" : ""}`;
      } else if (fpr) {
        from = `>=${from}`;
      } else {
        from = `>=${from}${incPr ? "-0" : ""}`;
      }
      if (isX(tM)) {
        to = "";
      } else if (isX(tm)) {
        to = `<${+tM + 1}.0.0-0`;
      } else if (isX(tp)) {
        to = `<${tM}.${+tm + 1}.0-0`;
      } else if (tpr) {
        to = `<=${tM}.${tm}.${tp}-${tpr}`;
      } else if (incPr) {
        to = `<${tM}.${tm}.${+tp + 1}-0`;
      } else {
        to = `<=${to}`;
      }
      return `${from} ${to}`.trim();
    };
    var testSet = (set, version, options) => {
      for (let i = 0; i < set.length; i++) {
        if (!set[i].test(version)) {
          return false;
        }
      }
      if (version.prerelease.length && !options.includePrerelease) {
        for (let i = 0; i < set.length; i++) {
          debug(set[i].semver);
          if (set[i].semver === Comparator.ANY) {
            continue;
          }
          if (set[i].semver.prerelease.length > 0) {
            const allowed = set[i].semver;
            if (allowed.major === version.major && allowed.minor === version.minor && allowed.patch === version.patch) {
              return true;
            }
          }
        }
        return false;
      }
      return true;
    };
  }
});

// ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/satisfies.js
var require_satisfies = __commonJS({
  "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/satisfies.js"(exports, module) {
    init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
    var Range = require_range();
    var satisfies = (version, range, options) => {
      try {
        range = new Range(range, options);
      } catch (er) {
        return false;
      }
      return range.test(version);
    };
    module.exports = satisfies;
  }
});

// src/pool/index.ts
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
import assert4 from "node:assert";
import crypto from "node:crypto";
import events from "node:events";
import fs3 from "node:fs";
import path5 from "node:path";
import { fileURLToPath as fileURLToPath2, pathToFileURL as pathToFileURL2 } from "node:url";
import util2 from "node:util";

// ../../node_modules/.pnpm/birpc@0.2.14/node_modules/birpc/dist/index.mjs
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
var DEFAULT_TIMEOUT = 6e4;
function defaultSerialize(i) {
  return i;
}
var defaultDeserialize = defaultSerialize;
var { clearTimeout, setTimeout: setTimeout2 } = globalThis;
var random = Math.random.bind(Math);
function createBirpc(functions, options) {
  const {
    post,
    on,
    eventNames = [],
    serialize = defaultSerialize,
    deserialize = defaultDeserialize,
    resolver,
    timeout: timeout2 = DEFAULT_TIMEOUT
  } = options;
  const rpcPromiseMap = /* @__PURE__ */ new Map();
  let _promise;
  const rpc = new Proxy({}, {
    get(_, method) {
      if (method === "$functions")
        return functions;
      const sendEvent = (...args) => {
        post(serialize({ m: method, a: args, t: "q" }));
      };
      if (eventNames.includes(method)) {
        sendEvent.asEvent = sendEvent;
        return sendEvent;
      }
      const sendCall = async (...args) => {
        await _promise;
        return new Promise((resolve2, reject) => {
          const id = nanoid();
          let timeoutId;
          if (timeout2 >= 0) {
            timeoutId = setTimeout2(() => {
              reject(new Error(`[birpc] timeout on calling "${method}"`));
              rpcPromiseMap.delete(id);
            }, timeout2).unref?.();
          }
          rpcPromiseMap.set(id, { resolve: resolve2, reject, timeoutId });
          post(serialize({ m: method, a: args, i: id, t: "q" }));
        });
      };
      sendCall.asEvent = sendEvent;
      return sendCall;
    }
  });
  _promise = on(async (data, ...extra) => {
    const msg = deserialize(data);
    if (msg.t === "q") {
      const { m: method, a: args } = msg;
      let result, error;
      const fn = resolver ? resolver(method, functions[method]) : functions[method];
      if (!fn) {
        error = new Error(`[birpc] function "${method}" not found`);
      } else {
        try {
          result = await fn.apply(rpc, args);
        } catch (e) {
          error = e;
        }
      }
      if (msg.i) {
        if (error && options.onError)
          options.onError(error, method, args);
        post(serialize({ t: "s", i: msg.i, r: result, e: error }), ...extra);
      }
    } else {
      const { i: ack, r: result, e: error } = msg;
      const promise = rpcPromiseMap.get(ack);
      if (promise) {
        clearTimeout(promise.timeoutId);
        if (error)
          promise.reject(error);
        else
          promise.resolve(result);
      }
      rpcPromiseMap.delete(ack);
    }
  });
  return rpc;
}
var urlAlphabet = "useandom-26T198340PX75pxJACKVERYMINDBUSHWOLF_GQZbfghjklqvwyzrict";
function nanoid(size = 21) {
  let id = "";
  let i = size;
  while (i--)
    id += urlAlphabet[random() * 64 | 0];
  return id;
}

// ../../node_modules/.pnpm/devalue@5.3.2/node_modules/devalue/index.js
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();

// ../../node_modules/.pnpm/devalue@5.3.2/node_modules/devalue/src/utils.js
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
var DevalueError = class extends Error {
  /**
   * @param {string} message
   * @param {string[]} keys
   */
  constructor(message, keys) {
    super(message);
    this.name = "DevalueError";
    this.path = keys.join("");
  }
};
function is_primitive(thing) {
  return Object(thing) !== thing;
}
var object_proto_names = /* @__PURE__ */ Object.getOwnPropertyNames(
  Object.prototype
).sort().join("\0");
function is_plain_object(thing) {
  const proto = Object.getPrototypeOf(thing);
  return proto === Object.prototype || proto === null || Object.getPrototypeOf(proto) === null || Object.getOwnPropertyNames(proto).sort().join("\0") === object_proto_names;
}
function get_type(thing) {
  return Object.prototype.toString.call(thing).slice(8, -1);
}
function get_escaped_char(char) {
  switch (char) {
    case '"':
      return '\\"';
    case "<":
      return "\\u003C";
    case "\\":
      return "\\\\";
    case "\n":
      return "\\n";
    case "\r":
      return "\\r";
    case "	":
      return "\\t";
    case "\b":
      return "\\b";
    case "\f":
      return "\\f";
    case "\u2028":
      return "\\u2028";
    case "\u2029":
      return "\\u2029";
    default:
      return char < " " ? `\\u${char.charCodeAt(0).toString(16).padStart(4, "0")}` : "";
  }
}
function stringify_string(str) {
  let result = "";
  let last_pos = 0;
  const len = str.length;
  for (let i = 0; i < len; i += 1) {
    const char = str[i];
    const replacement = get_escaped_char(char);
    if (replacement) {
      result += str.slice(last_pos, i) + replacement;
      last_pos = i + 1;
    }
  }
  return `"${last_pos === 0 ? str : result + str.slice(last_pos)}"`;
}
function enumerable_symbols(object) {
  return Object.getOwnPropertySymbols(object).filter(
    (symbol) => Object.getOwnPropertyDescriptor(object, symbol).enumerable
  );
}
var is_identifier = /^[a-zA-Z_$][a-zA-Z_$0-9]*$/;
function stringify_key(key) {
  return is_identifier.test(key) ? "." + key : "[" + JSON.stringify(key) + "]";
}

// ../../node_modules/.pnpm/devalue@5.3.2/node_modules/devalue/src/parse.js
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();

// ../../node_modules/.pnpm/devalue@5.3.2/node_modules/devalue/src/base64.js
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
function encode64(arraybuffer) {
  const dv = new DataView(arraybuffer);
  let binaryString = "";
  for (let i = 0; i < arraybuffer.byteLength; i++) {
    binaryString += String.fromCharCode(dv.getUint8(i));
  }
  return binaryToAscii(binaryString);
}
function decode64(string) {
  const binaryString = asciiToBinary(string);
  const arraybuffer = new ArrayBuffer(binaryString.length);
  const dv = new DataView(arraybuffer);
  for (let i = 0; i < arraybuffer.byteLength; i++) {
    dv.setUint8(i, binaryString.charCodeAt(i));
  }
  return arraybuffer;
}
var KEY_STRING = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
function asciiToBinary(data) {
  if (data.length % 4 === 0) {
    data = data.replace(/==?$/, "");
  }
  let output = "";
  let buffer = 0;
  let accumulatedBits = 0;
  for (let i = 0; i < data.length; i++) {
    buffer <<= 6;
    buffer |= KEY_STRING.indexOf(data[i]);
    accumulatedBits += 6;
    if (accumulatedBits === 24) {
      output += String.fromCharCode((buffer & 16711680) >> 16);
      output += String.fromCharCode((buffer & 65280) >> 8);
      output += String.fromCharCode(buffer & 255);
      buffer = accumulatedBits = 0;
    }
  }
  if (accumulatedBits === 12) {
    buffer >>= 4;
    output += String.fromCharCode(buffer);
  } else if (accumulatedBits === 18) {
    buffer >>= 2;
    output += String.fromCharCode((buffer & 65280) >> 8);
    output += String.fromCharCode(buffer & 255);
  }
  return output;
}
function binaryToAscii(str) {
  let out = "";
  for (let i = 0; i < str.length; i += 3) {
    const groupsOfSix = [void 0, void 0, void 0, void 0];
    groupsOfSix[0] = str.charCodeAt(i) >> 2;
    groupsOfSix[1] = (str.charCodeAt(i) & 3) << 4;
    if (str.length > i + 1) {
      groupsOfSix[1] |= str.charCodeAt(i + 1) >> 4;
      groupsOfSix[2] = (str.charCodeAt(i + 1) & 15) << 2;
    }
    if (str.length > i + 2) {
      groupsOfSix[2] |= str.charCodeAt(i + 2) >> 6;
      groupsOfSix[3] = str.charCodeAt(i + 2) & 63;
    }
    for (let j = 0; j < groupsOfSix.length; j++) {
      if (typeof groupsOfSix[j] === "undefined") {
        out += "=";
      } else {
        out += KEY_STRING[groupsOfSix[j]];
      }
    }
  }
  return out;
}

// ../../node_modules/.pnpm/devalue@5.3.2/node_modules/devalue/src/constants.js
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
var UNDEFINED = -1;
var HOLE = -2;
var NAN = -3;
var POSITIVE_INFINITY = -4;
var NEGATIVE_INFINITY = -5;
var NEGATIVE_ZERO = -6;

// ../../node_modules/.pnpm/devalue@5.3.2/node_modules/devalue/src/parse.js
function parse(serialized, revivers) {
  return unflatten(JSON.parse(serialized), revivers);
}
function unflatten(parsed, revivers) {
  if (typeof parsed === "number") return hydrate(parsed, true);
  if (!Array.isArray(parsed) || parsed.length === 0) {
    throw new Error("Invalid input");
  }
  const values = (
    /** @type {any[]} */
    parsed
  );
  const hydrated = Array(values.length);
  function hydrate(index, standalone = false) {
    if (index === UNDEFINED) return void 0;
    if (index === NAN) return NaN;
    if (index === POSITIVE_INFINITY) return Infinity;
    if (index === NEGATIVE_INFINITY) return -Infinity;
    if (index === NEGATIVE_ZERO) return -0;
    if (standalone || typeof index !== "number") {
      throw new Error(`Invalid input`);
    }
    if (index in hydrated) return hydrated[index];
    const value = values[index];
    if (!value || typeof value !== "object") {
      hydrated[index] = value;
    } else if (Array.isArray(value)) {
      if (typeof value[0] === "string") {
        const type = value[0];
        const reviver = revivers?.[type];
        if (reviver) {
          return hydrated[index] = reviver(hydrate(value[1]));
        }
        switch (type) {
          case "Date":
            hydrated[index] = new Date(value[1]);
            break;
          case "Set":
            const set = /* @__PURE__ */ new Set();
            hydrated[index] = set;
            for (let i = 1; i < value.length; i += 1) {
              set.add(hydrate(value[i]));
            }
            break;
          case "Map":
            const map = /* @__PURE__ */ new Map();
            hydrated[index] = map;
            for (let i = 1; i < value.length; i += 2) {
              map.set(hydrate(value[i]), hydrate(value[i + 1]));
            }
            break;
          case "RegExp":
            hydrated[index] = new RegExp(value[1], value[2]);
            break;
          case "Object":
            hydrated[index] = Object(value[1]);
            break;
          case "BigInt":
            hydrated[index] = BigInt(value[1]);
            break;
          case "null":
            const obj = /* @__PURE__ */ Object.create(null);
            hydrated[index] = obj;
            for (let i = 1; i < value.length; i += 2) {
              obj[value[i]] = hydrate(value[i + 1]);
            }
            break;
          case "Int8Array":
          case "Uint8Array":
          case "Uint8ClampedArray":
          case "Int16Array":
          case "Uint16Array":
          case "Int32Array":
          case "Uint32Array":
          case "Float32Array":
          case "Float64Array":
          case "BigInt64Array":
          case "BigUint64Array": {
            const TypedArrayConstructor = globalThis[type];
            const typedArray = new TypedArrayConstructor(hydrate(value[1]));
            hydrated[index] = value[2] !== void 0 ? typedArray.subarray(value[2], value[3]) : typedArray;
            break;
          }
          case "ArrayBuffer": {
            const base64 = value[1];
            const arraybuffer = decode64(base64);
            hydrated[index] = arraybuffer;
            break;
          }
          case "Temporal.Duration":
          case "Temporal.Instant":
          case "Temporal.PlainDate":
          case "Temporal.PlainTime":
          case "Temporal.PlainDateTime":
          case "Temporal.PlainMonthDay":
          case "Temporal.PlainYearMonth":
          case "Temporal.ZonedDateTime": {
            const temporalName = type.slice(9);
            hydrated[index] = Temporal[temporalName].from(value[1]);
            break;
          }
          case "URL": {
            const url = new URL(value[1]);
            hydrated[index] = url;
            break;
          }
          case "URLSearchParams": {
            const url = new URLSearchParams(value[1]);
            hydrated[index] = url;
            break;
          }
          default:
            throw new Error(`Unknown type ${type}`);
        }
      } else {
        const array = new Array(value.length);
        hydrated[index] = array;
        for (let i = 0; i < value.length; i += 1) {
          const n = value[i];
          if (n === HOLE) continue;
          array[i] = hydrate(n);
        }
      }
    } else {
      const object = {};
      hydrated[index] = object;
      for (const key in value) {
        if (key === "__proto__") {
          throw new Error("Cannot parse an object with a `__proto__` property");
        }
        const n = value[key];
        object[key] = hydrate(n);
      }
    }
    return hydrated[index];
  }
  return hydrate(0);
}

// ../../node_modules/.pnpm/devalue@5.3.2/node_modules/devalue/src/stringify.js
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
function stringify(value, reducers) {
  const stringified = [];
  const indexes = /* @__PURE__ */ new Map();
  const custom = [];
  if (reducers) {
    for (const key of Object.getOwnPropertyNames(reducers)) {
      custom.push({ key, fn: reducers[key] });
    }
  }
  const keys = [];
  let p = 0;
  function flatten(thing) {
    if (typeof thing === "function") {
      throw new DevalueError(`Cannot stringify a function`, keys);
    }
    if (thing === void 0) return UNDEFINED;
    if (Number.isNaN(thing)) return NAN;
    if (thing === Infinity) return POSITIVE_INFINITY;
    if (thing === -Infinity) return NEGATIVE_INFINITY;
    if (thing === 0 && 1 / thing < 0) return NEGATIVE_ZERO;
    if (indexes.has(thing)) return indexes.get(thing);
    const index2 = p++;
    indexes.set(thing, index2);
    for (const { key, fn } of custom) {
      const value2 = fn(thing);
      if (value2) {
        stringified[index2] = `["${key}",${flatten(value2)}]`;
        return index2;
      }
    }
    let str = "";
    if (is_primitive(thing)) {
      str = stringify_primitive(thing);
    } else {
      const type = get_type(thing);
      switch (type) {
        case "Number":
        case "String":
        case "Boolean":
          str = `["Object",${stringify_primitive(thing)}]`;
          break;
        case "BigInt":
          str = `["BigInt",${thing}]`;
          break;
        case "Date":
          const valid = !isNaN(thing.getDate());
          str = `["Date","${valid ? thing.toISOString() : ""}"]`;
          break;
        case "URL":
          str = `["URL",${stringify_string(thing.toString())}]`;
          break;
        case "URLSearchParams":
          str = `["URLSearchParams",${stringify_string(thing.toString())}]`;
          break;
        case "RegExp":
          const { source, flags } = thing;
          str = flags ? `["RegExp",${stringify_string(source)},"${flags}"]` : `["RegExp",${stringify_string(source)}]`;
          break;
        case "Array":
          str = "[";
          for (let i = 0; i < thing.length; i += 1) {
            if (i > 0) str += ",";
            if (i in thing) {
              keys.push(`[${i}]`);
              str += flatten(thing[i]);
              keys.pop();
            } else {
              str += HOLE;
            }
          }
          str += "]";
          break;
        case "Set":
          str = '["Set"';
          for (const value2 of thing) {
            str += `,${flatten(value2)}`;
          }
          str += "]";
          break;
        case "Map":
          str = '["Map"';
          for (const [key, value2] of thing) {
            keys.push(
              `.get(${is_primitive(key) ? stringify_primitive(key) : "..."})`
            );
            str += `,${flatten(key)},${flatten(value2)}`;
            keys.pop();
          }
          str += "]";
          break;
        case "Int8Array":
        case "Uint8Array":
        case "Uint8ClampedArray":
        case "Int16Array":
        case "Uint16Array":
        case "Int32Array":
        case "Uint32Array":
        case "Float32Array":
        case "Float64Array":
        case "BigInt64Array":
        case "BigUint64Array": {
          const typedArray = thing;
          str = '["' + type + '",' + flatten(typedArray.buffer);
          const a = thing.byteOffset;
          const b = a + thing.byteLength;
          if (a > 0 || b !== typedArray.buffer.byteLength) {
            const m = +/(\d+)/.exec(type)[1] / 8;
            str += `,${a / m},${b / m}`;
          }
          str += "]";
          break;
        }
        case "ArrayBuffer": {
          const arraybuffer = thing;
          const base64 = encode64(arraybuffer);
          str = `["ArrayBuffer","${base64}"]`;
          break;
        }
        case "Temporal.Duration":
        case "Temporal.Instant":
        case "Temporal.PlainDate":
        case "Temporal.PlainTime":
        case "Temporal.PlainDateTime":
        case "Temporal.PlainMonthDay":
        case "Temporal.PlainYearMonth":
        case "Temporal.ZonedDateTime":
          str = `["${type}",${stringify_string(thing.toString())}]`;
          break;
        default:
          if (!is_plain_object(thing)) {
            throw new DevalueError(
              `Cannot stringify arbitrary non-POJOs`,
              keys
            );
          }
          if (enumerable_symbols(thing).length > 0) {
            throw new DevalueError(
              `Cannot stringify POJOs with symbolic keys`,
              keys
            );
          }
          if (Object.getPrototypeOf(thing) === null) {
            str = '["null"';
            for (const key in thing) {
              keys.push(stringify_key(key));
              str += `,${stringify_string(key)},${flatten(thing[key])}`;
              keys.pop();
            }
            str += "]";
          } else {
            str = "{";
            let started = false;
            for (const key in thing) {
              if (started) str += ",";
              started = true;
              keys.push(stringify_key(key));
              str += `${stringify_string(key)}:${flatten(thing[key])}`;
              keys.pop();
            }
            str += "}";
          }
      }
    }
    stringified[index2] = str;
    return index2;
  }
  const index = flatten(value);
  if (index < 0) return `${index}`;
  return `[${stringified.join(",")}]`;
}
function stringify_primitive(thing) {
  const type = typeof thing;
  if (type === "string") return stringify_string(thing);
  if (thing instanceof String) return stringify_string(thing.toString());
  if (thing === void 0) return UNDEFINED.toString();
  if (thing === 0 && 1 / thing < 0) return NEGATIVE_ZERO.toString();
  if (type === "bigint") return `["BigInt","${thing}"]`;
  return String(thing);
}

// ../../node_modules/.pnpm/get-port@7.1.0/node_modules/get-port/index.js
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
import net from "node:net";
import os from "node:os";
var Locked = class extends Error {
  constructor(port) {
    super(`${port} is locked`);
  }
};
var lockedPorts = {
  old: /* @__PURE__ */ new Set(),
  young: /* @__PURE__ */ new Set()
};
var releaseOldLockedPortsIntervalMs = 1e3 * 15;
var minPort = 1024;
var maxPort = 65535;
var timeout;
var getLocalHosts = () => {
  const interfaces = os.networkInterfaces();
  const results = /* @__PURE__ */ new Set([void 0, "0.0.0.0"]);
  for (const _interface of Object.values(interfaces)) {
    for (const config of _interface) {
      results.add(config.address);
    }
  }
  return results;
};
var checkAvailablePort = (options) => new Promise((resolve2, reject) => {
  const server = net.createServer();
  server.unref();
  server.on("error", reject);
  server.listen(options, () => {
    const { port } = server.address();
    server.close(() => {
      resolve2(port);
    });
  });
});
var getAvailablePort = async (options, hosts) => {
  if (options.host || options.port === 0) {
    return checkAvailablePort(options);
  }
  for (const host of hosts) {
    try {
      await checkAvailablePort({ port: options.port, host });
    } catch (error) {
      if (!["EADDRNOTAVAIL", "EINVAL"].includes(error.code)) {
        throw error;
      }
    }
  }
  return options.port;
};
var portCheckSequence = function* (ports) {
  if (ports) {
    yield* ports;
  }
  yield 0;
};
async function getPorts(options) {
  let ports;
  let exclude = /* @__PURE__ */ new Set();
  if (options) {
    if (options.port) {
      ports = typeof options.port === "number" ? [options.port] : options.port;
    }
    if (options.exclude) {
      const excludeIterable = options.exclude;
      if (typeof excludeIterable[Symbol.iterator] !== "function") {
        throw new TypeError("The `exclude` option must be an iterable.");
      }
      for (const element of excludeIterable) {
        if (typeof element !== "number") {
          throw new TypeError("Each item in the `exclude` option must be a number corresponding to the port you want excluded.");
        }
        if (!Number.isSafeInteger(element)) {
          throw new TypeError(`Number ${element} in the exclude option is not a safe integer and can't be used`);
        }
      }
      exclude = new Set(excludeIterable);
    }
  }
  if (timeout === void 0) {
    timeout = setTimeout(() => {
      timeout = void 0;
      lockedPorts.old = lockedPorts.young;
      lockedPorts.young = /* @__PURE__ */ new Set();
    }, releaseOldLockedPortsIntervalMs);
    if (timeout.unref) {
      timeout.unref();
    }
  }
  const hosts = getLocalHosts();
  for (const port of portCheckSequence(ports)) {
    try {
      if (exclude.has(port)) {
        continue;
      }
      let availablePort = await getAvailablePort({ ...options, port }, hosts);
      while (lockedPorts.old.has(availablePort) || lockedPorts.young.has(availablePort)) {
        if (port !== 0) {
          throw new Locked(port);
        }
        availablePort = await getAvailablePort({ ...options, port }, hosts);
      }
      lockedPorts.young.add(availablePort);
      return availablePort;
    } catch (error) {
      if (!["EADDRINUSE", "EACCES"].includes(error.code) && !(error instanceof Locked)) {
        throw error;
      }
    }
  }
  throw new Error("No available ports found");
}
function portNumbers(from, to) {
  if (!Number.isInteger(from) || !Number.isInteger(to)) {
    throw new TypeError("`from` and `to` must be integer numbers");
  }
  if (from < minPort || from > maxPort) {
    throw new RangeError(`'from' must be between ${minPort} and ${maxPort}`);
  }
  if (to < minPort || to > maxPort) {
    throw new RangeError(`'to' must be between ${minPort} and ${maxPort}`);
  }
  if (from > to) {
    throw new RangeError("`to` must be greater than or equal to `from`");
  }
  const generator = function* (from2, to2) {
    for (let port = from2; port <= to2; port++) {
      yield port;
    }
  };
  return generator(from, to);
}

// src/pool/index.ts
var import_satisfies = __toESM(require_satisfies(), 1);
import {
  compileModuleRules,
  getNodeCompat,
  kCurrentWorker,
  kUnsafeEphemeralUniqueKey,
  Log as Log2,
  LogLevel as LogLevel2,
  maybeApply,
  Miniflare,
  structuredSerializableReducers,
  structuredSerializableRevivers,
  supportedCompatibilityDate,
  testRegExps,
  WebSocket
} from "miniflare";
import { createMethodsRPC } from "vitest/node";
import { experimental_readRawConfig } from "wrangler";

// src/shared/builtin-modules.ts
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
var workerdBuiltinModules = /* @__PURE__ */ new Set([
  ...define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES_default,
  "__STATIC_CONTENT_MANIFEST"
]);

// src/shared/chunking-socket.ts
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
import assert from "node:assert";
import { Buffer as Buffer2 } from "node:buffer";
function createChunkingSocket(socket, maxChunkByteLength = 1048576) {
  const listeners = [];
  const decoder = new TextDecoder();
  let chunks;
  socket.on((message) => {
    if (typeof message === "string") {
      if (chunks !== void 0) {
        assert.strictEqual(message, "", "Expected end-of-chunks");
        message = chunks + decoder.decode();
        chunks = void 0;
      }
      for (const listener of listeners) {
        listener(message);
      }
    } else {
      chunks ??= "";
      chunks += decoder.decode(message, { stream: true });
    }
  });
  return {
    post(value) {
      if (Buffer2.byteLength(value) > maxChunkByteLength) {
        const encoded = Buffer2.from(value);
        for (let i = 0; i < encoded.byteLength; i += maxChunkByteLength) {
          socket.post(encoded.subarray(i, i + maxChunkByteLength));
        }
        socket.post("");
      } else {
        socket.post(value);
      }
    },
    on(listener) {
      listeners.push(listener);
    }
  };
}

// src/pool/compatibility-flag-assertions.ts
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
var CompatibilityFlagAssertions = class {
  #compatibilityDate;
  #compatibilityFlags;
  #optionsPath;
  #relativeProjectPath;
  #relativeWranglerConfigPath;
  constructor(options) {
    this.#compatibilityDate = options.compatibilityDate;
    this.#compatibilityFlags = options.compatibilityFlags;
    this.#optionsPath = options.optionsPath;
    this.#relativeProjectPath = options.relativeProjectPath;
    this.#relativeWranglerConfigPath = options.relativeWranglerConfigPath;
  }
  /**
   * Checks if a specific flag is present in the compatibilityFlags array.
   */
  #flagExists(flag) {
    return this.#compatibilityFlags.includes(flag);
  }
  /**
   * Constructs the base of the error message.
   *
   * @example
   * In project /path/to/project
   *
   * @example
   * In project /path/to/project's configuration file wrangler.toml
   */
  #buildErrorMessageBase() {
    let message = `In project ${this.#relativeProjectPath}`;
    if (this.#relativeWranglerConfigPath) {
      message += `'s configuration file ${this.#relativeWranglerConfigPath}`;
    }
    return message;
  }
  /**
   * Constructs the configuration path part of the error message.
   */
  #buildConfigPath(setting) {
    if (this.#relativeWranglerConfigPath) {
      return `\`${setting}\``;
    }
    const camelCaseSetting = setting.replace(
      /_(\w)/g,
      (_, letter) => letter.toUpperCase()
    );
    return `\`${this.#optionsPath}.${camelCaseSetting}\``;
  }
  isEnabled(enableFlag, disableFlag, defaultOnDate) {
    return !this.#flagExists(disableFlag) || this.#flagExists(enableFlag) || isDateSufficient(this.#compatibilityDate, defaultOnDate);
  }
  /**
   * Ensures that a specific enable flag is present or that the compatibility date meets the required date.
   */
  assertIsEnabled({
    enableFlag,
    disableFlag,
    defaultOnDate
  }) {
    if (this.#flagExists(disableFlag)) {
      const errorMessage = `${this.#buildErrorMessageBase()}, ${this.#buildConfigPath(
        "compatibility_flags"
      )} must not contain "${disableFlag}".
This flag is incompatible with \`@cloudflare/vitest-pool-workers\`.`;
      return { isValid: false, errorMessage };
    }
    const enableFlagPresent = this.#flagExists(enableFlag);
    const dateSufficient = isDateSufficient(
      this.#compatibilityDate,
      defaultOnDate
    );
    if (!enableFlagPresent && !dateSufficient) {
      let errorMessage = `${this.#buildErrorMessageBase()}, ${this.#buildConfigPath(
        "compatibility_flags"
      )} must contain "${enableFlag}"`;
      if (defaultOnDate) {
        errorMessage += `, or ${this.#buildConfigPath(
          "compatibility_date"
        )} must be >= "${defaultOnDate}".`;
      }
      errorMessage += `
This flag is required to use \`@cloudflare/vitest-pool-workers\`.`;
      return { isValid: false, errorMessage };
    }
    return { isValid: true };
  }
  /**
   * Ensures that a any one of a given set of flags is present in the compatibility_flags array.
   */
  assertAtLeastOneFlagExists(flags) {
    if (flags.length === 0 || flags.some((flag) => this.#flagExists(flag))) {
      return { isValid: true };
    }
    const errorMessage = `${this.#buildErrorMessageBase()}, ${this.#buildConfigPath(
      "compatibility_flags"
    )} must contain one of ${flags.map((flag) => `"${flag}"`).join("/")}.
Either one of these flags is required to use \`@cloudflare/vitest-pool-workers\`.`;
    return { isValid: false, errorMessage };
  }
};
function parseDate(dateStr) {
  const date = new Date(dateStr);
  if (isNaN(date.getTime())) {
    throw new Error(`Invalid date format: "${dateStr}"`);
  }
  return date;
}
function isDateSufficient(compatibilityDate, defaultOnDate) {
  if (!compatibilityDate || !defaultOnDate) {
    return false;
  }
  const compDate = parseDate(compatibilityDate);
  const reqDate = parseDate(defaultOnDate);
  return compDate >= reqDate;
}

// src/pool/config.ts
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
import path2 from "node:path";
import {
  formatZodError,
  getRootPath,
  Log,
  LogLevel,
  mergeWorkerOptions,
  parseWithRootPath,
  PLUGINS
} from "miniflare";
import { z } from "zod";

// src/pool/helpers.ts
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
import path from "node:path";
var WORKER_NAME_PREFIX = "vitest-pool-workers-";
function isFileNotFoundError(e) {
  return typeof e === "object" && e !== null && "code" in e && e.code === "ENOENT";
}
function getProjectPath(project) {
  return project.config.config ?? project.path;
}
function getRelativeProjectPath(project) {
  const projectPath = getProjectPath(project);
  if (typeof projectPath === "number") {
    return projectPath;
  } else {
    return path.relative("", projectPath);
  }
}

// src/pool/config.ts
var PLUGIN_VALUES = Object.values(PLUGINS);
var OPTIONS_PATH_ARRAY = ["test", "poolOptions", "workers"];
var OPTIONS_PATH = OPTIONS_PATH_ARRAY.join(".");
var WorkersPoolOptionsSchema = z.object({
  /**
   * Entrypoint to Worker run in the same isolate/context as tests. This is
   * required to use `import { SELF } from "cloudflare:test"`, or Durable
   * Objects without an explicit `scriptName`. Note this goes through Vite
   * transforms and can be a TypeScript file. Note also
   * `import module from "<path-to-main>"` inside tests gives exactly the same
   * `module` instance as is used internally for the `SELF` and Durable Object
   * bindings.
   */
  main: z.ostring(),
  /**
   * Enables per-test isolated storage. If enabled, any writes to storage
   * performed in a test will be undone at the end of the test. The test storage
   * environment is copied from the containing suite, meaning `beforeAll()`
   * hooks can be used to seed data. If this is disabled, all tests will share
   * the same storage.
   */
  isolatedStorage: z.boolean().default(true),
  /**
   * Enables remote bindings to access remote resources configured
   * with `remote: true` in the wrangler configuration file.
   */
  remoteBindings: z.boolean().default(true),
  /**
   * Additional exports.
   * A map of module exports to be made available on the `ctx.exports`
   * that cannot be automatically inferred by analyzing the Worker source code.
   *
   * This is useful for exports that are re-exported implicitly, for example
   * through wildcard (`export * from "..."`) re-exports from virtual modules.
   */
  additionalExports: z.record(
    z.string(),
    z.union([
      z.literal("WorkerEntrypoint"),
      z.literal("DurableObject"),
      z.literal("WorkflowEntrypoint")
    ])
  ).default({}),
  /**
   * If enabled, Workers will be run in a single shared worker instance.
   */
  /**
   * Runs all tests in this project serially in the same worker, using the same
   * module cache. This can significantly speed up tests if you've got lots of
   * small test files.
   */
  singleWorker: z.boolean().default(false),
  miniflare: z.object({
    workers: z.array(z.object({}).passthrough()).optional()
  }).passthrough().optional(),
  wrangler: z.object({ configPath: z.ostring(), environment: z.ostring() }).optional()
});
function isZodErrorLike(value) {
  return typeof value === "object" && value !== null && "issues" in value && Array.isArray(value.issues);
}
function coalesceZodErrors(ref, thrown) {
  if (!isZodErrorLike(thrown)) {
    throw thrown;
  }
  if (ref.value === void 0) {
    ref.value = thrown;
  } else {
    ref.value.issues.push(...thrown.issues);
  }
}
function parseWorkerOptions(rootPath, value, withoutScript, opts) {
  if (withoutScript) {
    value["script"] = "";
    delete value["scriptPath"];
    delete value["modules"];
    delete value["modulesRoot"];
  }
  const result = {};
  const errorRef = {};
  for (const plugin of PLUGIN_VALUES) {
    try {
      const parsed = parseWithRootPath(rootPath, plugin.options, value, opts);
      Object.assign(result, parsed);
    } catch (e) {
      coalesceZodErrors(errorRef, e);
    }
  }
  if (errorRef.value !== void 0) {
    throw errorRef.value;
  }
  if (withoutScript) {
    delete value["script"];
  }
  return result;
}
var log = new Log(LogLevel.WARN, { prefix: "vpw" });
function filterTails(tails, userWorkers) {
  return tails?.filter((tailService) => {
    let name;
    if (typeof tailService === "string") {
      name = tailService;
    } else if (typeof tailService === "object" && "name" in tailService && typeof tailService.name === "string") {
      name = tailService.name;
    } else {
      return true;
    }
    const found = userWorkers?.some((w) => w.name === name);
    if (!found) {
      log.warn(
        `Tail consumer "${name}" was not found in your config. Make sure you add it if you'd like to simulate receiving tail events locally.`
      );
    }
    return found;
  });
}
var remoteProxySessionsDataMap = /* @__PURE__ */ new Map();
async function parseCustomPoolOptions(rootPath, value, opts) {
  const options = WorkersPoolOptionsSchema.parse(
    value,
    opts
  );
  options.miniflare ??= {};
  const errorRef = {};
  const workers = options.miniflare?.workers;
  const rootPathOption = getRootPath(options.miniflare);
  rootPath = path2.resolve(rootPath, rootPathOption);
  try {
    options.miniflare = parseWorkerOptions(
      rootPath,
      options.miniflare,
      /* withoutScript */
      true,
      // (script provided by runner)
      { path: [...opts.path, "miniflare"] }
    );
  } catch (e) {
    coalesceZodErrors(errorRef, e);
  }
  options.miniflare.workers = [];
  if (workers !== void 0) {
    options.miniflare.workers = workers.map((worker, i) => {
      try {
        const workerRootPathOption = getRootPath(worker);
        const workerRootPath = path2.resolve(rootPath, workerRootPathOption);
        return parseWorkerOptions(
          workerRootPath,
          worker,
          /* withoutScript */
          false,
          {
            path: [...opts.path, "miniflare", "workers", i]
          }
        );
      } catch (e) {
        coalesceZodErrors(errorRef, e);
        return { script: "" };
      }
    });
  }
  if (errorRef.value !== void 0) {
    throw errorRef.value;
  }
  if (options.wrangler?.configPath !== void 0) {
    const configPath = path2.resolve(rootPath, options.wrangler.configPath);
    options.wrangler.configPath = configPath;
    const wrangler = await import("wrangler");
    const preExistingRemoteProxySessionData = options.wrangler?.configPath ? remoteProxySessionsDataMap.get(options.wrangler.configPath) : void 0;
    const remoteProxySessionData = options.remoteBindings ?? true ? await wrangler.maybeStartOrUpdateRemoteProxySession(
      {
        path: options.wrangler.configPath,
        environment: options.wrangler.environment
      },
      preExistingRemoteProxySessionData ?? null
    ) : null;
    if (options.wrangler?.configPath && remoteProxySessionData) {
      remoteProxySessionsDataMap.set(
        options.wrangler.configPath,
        remoteProxySessionData
      );
    }
    const { workerOptions, externalWorkers, define, main } = wrangler.unstable_getMiniflareWorkerOptions(
      configPath,
      options.wrangler.environment,
      {
        overrides: {
          assets: options.miniflare.assets,
          // doesn't work with containers yet so let's just disable it
          enableContainers: false
        },
        remoteProxyConnectionString: remoteProxySessionData?.session?.remoteProxyConnectionString
      }
    );
    const wrappedBindings = Object.values(workerOptions.wrappedBindings ?? {});
    const hasAIOrVectorizeBindings = wrappedBindings.some((binding) => {
      return typeof binding === "object" && (binding.scriptName.includes("__WRANGLER_EXTERNAL_VECTORIZE_WORKER") || binding.scriptName.includes("__WRANGLER_EXTERNAL_AI_WORKER"));
    });
    if (hasAIOrVectorizeBindings) {
      log.warn(
        "Workers AI and Vectorize bindings will access your Cloudflare account and incur usage charges even in testing. We recommend mocking any usage of these bindings in your tests."
      );
    }
    options.main ??= main;
    options.miniflare.workers = [
      ...options.miniflare.workers,
      ...externalWorkers
    ];
    options.miniflare = mergeWorkerOptions(
      workerOptions,
      options.miniflare
    );
    options.miniflare = {
      ...options.miniflare,
      tails: filterTails(workerOptions.tails, options.miniflare.workers)
    };
    options.defines = define;
  }
  if (options.miniflare?.assets) {
    options.miniflare.hasAssetsAndIsVitest = true;
    options.miniflare.assets.routerConfig ??= {};
    options.miniflare.assets.routerConfig.has_user_worker = Boolean(
      options.main
    );
  }
  return options;
}
async function parseProjectOptions(project) {
  const environment = project.config.environment;
  if (environment !== void 0 && environment !== "node") {
    const quotedEnvironment = JSON.stringify(environment);
    let migrationGuide = ".";
    if (environment === "miniflare") {
      migrationGuide = ", and refer to the migration guide if upgrading from `vitest-environment-miniflare`:\nhttps://developers.cloudflare.com/workers/testing/vitest-integration/get-started/migrate-from-miniflare-2/";
    }
    const relativePath = getRelativeProjectPath(project);
    const message = [
      `Unexpected custom \`environment\` ${quotedEnvironment} in project ${relativePath}.`,
      "The Workers pool always runs your tests inside of an environment providing Workers runtime APIs.",
      `Please remove the \`environment\` configuration${migrationGuide}`,
      "Use `poolMatchGlobs`/`environmentMatchGlobs` to run a subset of your tests in a different pool/environment."
    ].join("\n");
    throw new TypeError(message);
  }
  const projectPath = getProjectPath(project);
  const rootPath = typeof projectPath === "string" ? path2.dirname(projectPath) : "";
  const poolOptions = project.config.poolOptions;
  let workersPoolOptions = poolOptions?.workers ?? {};
  try {
    if (typeof workersPoolOptions === "function") {
      const inject = (key) => {
        return project.getProvidedContext()[key];
      };
      workersPoolOptions = await workersPoolOptions({ inject });
    }
    return await parseCustomPoolOptions(rootPath, workersPoolOptions, {
      path: OPTIONS_PATH_ARRAY
    });
  } catch (e) {
    if (!isZodErrorLike(e)) {
      throw e;
    }
    let formatted;
    try {
      formatted = formatZodError(e, {
        test: { poolOptions: { workers: workersPoolOptions } }
      });
    } catch {
      throw e;
    }
    const relativePath = getRelativeProjectPath(project);
    throw new TypeError(
      `Unexpected pool options in project ${relativePath}:
${formatted}`
    );
  }
}

// src/pool/guess-exports.ts
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
import path3 from "node:path";
import { build } from "esbuild";
async function guessWorkerExports(entryFile, additionalExports = {}) {
  const exportMap = new Map(Object.entries(additionalExports));
  if (entryFile === void 0) {
    return exportMap;
  }
  if (path3.extname(entryFile) == ".py") {
    return exportMap;
  }
  const result = await build({
    target: "es2024",
    loader: { ".js": "jsx", ".mjs": "jsx", ".cjs": "jsx" },
    entryPoints: [path3.resolve(entryFile)],
    metafile: true,
    bundle: false,
    write: false,
    logLevel: "silent"
  });
  const entryPoints = Object.values(result.metafile.outputs).filter(
    (output) => output.entryPoint !== void 0
  );
  if (entryPoints.length === 0) {
    throw new Error(
      `Cannot find entry-point "${entryFile}" in generated bundle.`
    );
  }
  if (entryPoints.length > 1) {
    throw new Error(
      "More than one entry-point found for generated bundle.\n" + entryPoints.map((output) => ` - ${output.entryPoint}`).join("\n")
    );
  }
  for (const exportName of entryPoints[0].exports) {
    if (!exportMap.has(exportName)) {
      exportMap.set(exportName, null);
    }
  }
  return exportMap;
}

// src/pool/loopback.ts
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
import assert2 from "node:assert";
import fs from "node:fs/promises";
import path4 from "node:path";
import {
  CACHE_PLUGIN_NAME,
  D1_PLUGIN_NAME,
  DURABLE_OBJECTS_PLUGIN_NAME,
  KV_PLUGIN_NAME,
  Mutex,
  R2_PLUGIN_NAME,
  Response,
  WORKFLOWS_PLUGIN_NAME
} from "miniflare";
async function handleSnapshotRequest(request, url) {
  const filePath = url.searchParams.get("path");
  if (filePath === null) {
    return new Response(null, { status: 400 });
  }
  if (request.method === "POST") {
    await fs.mkdir(filePath, { recursive: true });
    return new Response(null, { status: 204 });
  }
  if (request.method === "PUT") {
    const snapshot = await request.arrayBuffer();
    await fs.mkdir(path4.posix.dirname(filePath), { recursive: true });
    await fs.writeFile(filePath, new Uint8Array(snapshot));
    return new Response(null, { status: 204 });
  }
  if (request.method === "GET") {
    try {
      return new Response(await fs.readFile(filePath));
    } catch (e) {
      if (!isFileNotFoundError(e)) {
        throw e;
      }
      return new Response(null, { status: 404 });
    }
  }
  if (request.method === "DELETE") {
    try {
      await fs.unlink(filePath);
    } catch (e) {
      if (!isFileNotFoundError(e)) {
        throw e;
      }
    }
    return new Response(null, { status: 204 });
  }
  return new Response(null, { status: 405 });
}
async function emptyDir(dirPath) {
  let names;
  try {
    names = await fs.readdir(dirPath);
  } catch (e) {
    if (isFileNotFoundError(e)) {
      return;
    }
    throw e;
  }
  for (const name of names) {
    try {
      await fs.rm(path4.join(dirPath, name), { recursive: true, force: true });
    } catch (e) {
      if (isEbusyError(e)) {
        console.warn(
          `vitest-pool-worker: Unable to remove temporary directory: ${e}`
        );
      }
    }
  }
}
function isEbusyError(e) {
  return e instanceof Error && "code" in e && e.code === "EBUSY";
}
var stackStates = /* @__PURE__ */ new WeakMap();
function getState(mf) {
  let state = stackStates.get(mf);
  if (state === void 0) {
    const persistPaths = mf.unsafeGetPersistPaths();
    const durableObjectPersistPath = persistPaths.get("do");
    assert2(
      durableObjectPersistPath !== void 0,
      "Expected Durable Object persist path"
    );
    state = {
      mutex: new Mutex(),
      depth: 0,
      broken: false,
      persistPaths: Array.from(new Set(persistPaths.values())),
      durableObjectPersistPath
    };
    stackStates.set(mf, state);
  }
  return state;
}
var ABORT_ALL_WORKER_NAME = `${WORKER_NAME_PREFIX}abort-all`;
var ABORT_ALL_WORKER = {
  name: ABORT_ALL_WORKER_NAME,
  compatibilityFlags: ["unsafe_module"],
  modules: [
    {
      type: "ESModule",
      path: "index.mjs",
      contents: `
			import workerdUnsafe from "workerd:unsafe";
			export default {
				async fetch(request) {
					if (request.method !== "DELETE") return new Response(null, { status: 405 });
					await workerdUnsafe.abortAllDurableObjects();
					return new Response(null, { status: 204 });
				}
			};
			`
    }
  ]
};
function scheduleStorageReset(mf) {
  const state = getState(mf);
  assert2(state.storageResetPromise === void 0);
  state.storageResetPromise = state.mutex.runWith(async () => {
    const abortAllWorker = await mf.getWorker(ABORT_ALL_WORKER_NAME);
    await abortAllWorker.fetch("http://placeholder", { method: "DELETE" });
    for (const persistPath of state.persistPaths) {
      await emptyDir(persistPath);
    }
    state.depth = 0;
    state.storageResetPromise = void 0;
  });
}
async function waitForStorageReset(mf) {
  await getState(mf).storageResetPromise;
}
var BLOBS_DIR_NAME = "blobs";
var STACK_DIR_NAME = "__vitest_pool_workers_stack";
async function pushStackedStorage(intoDepth, persistPath) {
  const stackFramePath = path4.join(
    persistPath,
    STACK_DIR_NAME,
    intoDepth.toString()
  );
  await fs.mkdir(stackFramePath, { recursive: true });
  for (const key of await fs.readdir(persistPath, { withFileTypes: true })) {
    if (key.name === STACK_DIR_NAME) {
      continue;
    }
    const keyPath = path4.join(persistPath, key.name);
    const stackFrameKeyPath = path4.join(stackFramePath, key.name);
    assert2(key.isDirectory(), `Expected ${keyPath} to be a directory`);
    let createdStackFrameKeyPath = false;
    for (const name of await fs.readdir(keyPath)) {
      if (name === BLOBS_DIR_NAME) {
        break;
      }
      if (!createdStackFrameKeyPath) {
        createdStackFrameKeyPath = true;
        await fs.mkdir(stackFrameKeyPath);
      }
      const namePath = path4.join(keyPath, name);
      const stackFrameNamePath = path4.join(stackFrameKeyPath, name);
      assert2(name.endsWith(".sqlite"), `Expected .sqlite, got ${namePath}`);
      await fs.copyFile(namePath, stackFrameNamePath);
    }
  }
}
async function popStackedStorage(fromDepth, persistPath) {
  for (const key of await fs.readdir(persistPath, { withFileTypes: true })) {
    if (key.name === STACK_DIR_NAME) {
      continue;
    }
    const keyPath = path4.join(persistPath, key.name);
    for (const name of await fs.readdir(keyPath)) {
      if (name === BLOBS_DIR_NAME) {
        break;
      }
      const namePath = path4.join(keyPath, name);
      assert2(name.endsWith(".sqlite"), `Expected .sqlite, got ${namePath}`);
      await fs.unlink(namePath);
    }
  }
  const stackFramePath = path4.join(
    persistPath,
    STACK_DIR_NAME,
    fromDepth.toString()
  );
  await fs.cp(stackFramePath, persistPath, { recursive: true });
  await fs.rm(stackFramePath, { recursive: true, force: true });
}
var PLUGIN_PRODUCT_NAMES = {
  [CACHE_PLUGIN_NAME]: "Cache",
  [D1_PLUGIN_NAME]: "D1",
  [DURABLE_OBJECTS_PLUGIN_NAME]: "Durable Objects",
  [KV_PLUGIN_NAME]: "KV",
  [R2_PLUGIN_NAME]: "R2",
  [WORKFLOWS_PLUGIN_NAME]: "Workflows"
};
var LIST_FORMAT = new Intl.ListFormat("en-US");
function checkAllStorageOperationsResolved(action, source, persistPaths, results) {
  const failedProducts = [];
  const lines = [];
  for (let i = 0; i < results.length; i++) {
    const result = results[i];
    if (result.status === "rejected") {
      const pluginName = path4.basename(persistPaths[i]);
      const productName = PLUGIN_PRODUCT_NAMES[pluginName] ?? pluginName;
      failedProducts.push(productName);
      lines.push(`- ${result.reason}`);
    }
  }
  if (failedProducts.length > 0) {
    const separator = "=".repeat(80);
    lines.unshift(
      "",
      separator,
      `Failed to ${action} isolated storage stack frame in ${source}.`,
      `In particular, we were unable to ${action} ${LIST_FORMAT.format(failedProducts)} storage.`,
      "This usually means your Worker tried to access storage outside of a test, or some resources have not been disposed of properly.",
      `Ensure you "await" all Promises that read or write to these services, and make sure you use the "using" keyword when passing data across JSRPC.`,
      `See https://developers.cloudflare.com/workers/testing/vitest-integration/known-issues/#isolated-storage for more details.`,
      "\x1B[2m"
    );
    lines.push("\x1B[22m" + separator, "");
    if (failedProducts.includes(
      PLUGIN_PRODUCT_NAMES[WORKFLOWS_PLUGIN_NAME] ?? WORKFLOWS_PLUGIN_NAME
    )) {
      console.warn(
        [
          "",
          separator,
          `Workflows are being created in ${source}.`,
          "Even with isolated storage, Workflows are required to be manually disposed at the end of each test.",
          "See https://developers.cloudflare.com/workers/testing/vitest-integration/test-apis/ for more details.",
          ""
        ].join("\n")
      );
    }
    console.error(lines.join("\n"));
    return false;
  }
  return true;
}
async function handleStorageRequest(request, mf) {
  const state = getState(mf);
  if (state.broken) {
    return new Response(
      "Isolated storage failed. There should be additional logs above.",
      { status: 500 }
    );
  }
  const source = request.headers.get("MF-Vitest-Source") ?? "an unknown location";
  let success;
  if (request.method === "POST") {
    success = await state.mutex.runWith(async () => {
      state.depth++;
      const results = await Promise.allSettled(
        state.persistPaths.map(
          (persistPath) => pushStackedStorage(state.depth, persistPath)
        )
      );
      return checkAllStorageOperationsResolved(
        "push",
        source,
        state.persistPaths,
        results
      );
    });
  } else if (request.method === "DELETE") {
    success = await state.mutex.runWith(async () => {
      assert2(state.depth > 0, "Stack underflow");
      const results = await Promise.allSettled(
        state.persistPaths.map(
          (persistPath) => popStackedStorage(state.depth, persistPath)
        )
      );
      state.depth--;
      return checkAllStorageOperationsResolved(
        "pop",
        source,
        state.persistPaths,
        results
      );
    });
  } else {
    return new Response(null, { status: 405 });
  }
  if (success) {
    return new Response(null, { status: 204 });
  } else {
    state.broken = true;
    return new Response(
      "Isolated storage failed. There should be additional logs above.",
      { status: 500 }
    );
  }
}
async function handleDurableObjectsRequest(request, mf, url) {
  if (request.method !== "GET") {
    return new Response(null, { status: 405 });
  }
  const { durableObjectPersistPath } = getState(mf);
  const uniqueKey = url.searchParams.get("unique_key");
  if (uniqueKey === null) {
    return new Response(null, { status: 400 });
  }
  const namespacePath = path4.join(durableObjectPersistPath, uniqueKey);
  const ids = [];
  try {
    const names = await fs.readdir(namespacePath);
    for (const name of names) {
      if (name.endsWith(".sqlite")) {
        ids.push(name.substring(
          0,
          name.length - 7
          /* ".sqlite".length */
        ));
      }
    }
  } catch (e) {
    if (!isFileNotFoundError(e)) {
      throw e;
    }
  }
  return Response.json(ids);
}
function handleLoopbackRequest(request, mf) {
  const url = new URL(request.url);
  if (url.pathname === "/snapshot") {
    return handleSnapshotRequest(request, url);
  }
  if (url.pathname === "/storage") {
    return handleStorageRequest(request, mf);
  }
  if (url.pathname === "/durable-objects") {
    return handleDurableObjectsRequest(request, mf, url);
  }
  return new Response(null, { status: 404 });
}

// src/pool/module-fallback.ts
init_define_VITEST_POOL_WORKERS_DEFINE_BUILTIN_MODULES();
import assert3 from "node:assert";
import fs2 from "node:fs";
import { createRequire } from "node:module";
import platformPath from "node:path";
import posixPath from "node:path/posix";
import { fileURLToPath, pathToFileURL } from "node:url";
import util from "node:util";
import * as cjsModuleLexer from "cjs-module-lexer";
import { ModuleRuleTypeSchema, Response as Response2 } from "miniflare";
var debuglog = util.debuglog(
  "vitest-pool-workers:module-fallback",
  (log3) => debuglog = log3
);
var isWindows = process.platform === "win32";
function ensurePosixLikePath(filePath) {
  return isWindows ? filePath.replaceAll("\\", "/") : filePath;
}
var __filename = fileURLToPath(import.meta.url);
var __dirname = platformPath.dirname(__filename);
var require2 = createRequire(__filename);
var distPath = ensurePosixLikePath(platformPath.resolve(__dirname, ".."));
var libPath = posixPath.join(distPath, "worker", "lib");
var emptyLibPath = posixPath.join(libPath, "cloudflare/empty-internal.cjs");
var disableCjsEsmShimSuffix = "?mf_vitest_no_cjs_esm_shim";
function trimSuffix(suffix, value) {
  assert3(value.endsWith(suffix));
  return value.substring(0, value.length - suffix.length);
}
var versionHashRegExp = /\?v=[0-9a-f]+$/;
function trimViteVersionHash(filePath) {
  return filePath.replace(versionHashRegExp, "");
}
var forceModuleTypeRegexp = new RegExp(
  `\\?mf_vitest_force=(${ModuleRuleTypeSchema.options.join("|")})$`
);
function isFile(filePath) {
  try {
    return fs2.statSync(filePath).isFile();
  } catch (e) {
    if (isFileNotFoundError(e)) {
      return false;
    }
    throw e;
  }
}
function isDirectory(filePath) {
  try {
    return fs2.statSync(filePath).isDirectory();
  } catch (e) {
    if (isFileNotFoundError(e)) {
      return false;
    }
    throw e;
  }
}
function getParentPaths(filePath) {
  const parentPaths = [];
  while (true) {
    const parentPath = posixPath.dirname(filePath);
    if (parentPath === filePath) {
      return parentPaths;
    }
    parentPaths.push(parentPath);
    filePath = parentPath;
  }
}
var dirPathTypeModuleCache = /* @__PURE__ */ new Map();
function isWithinTypeModuleContext(filePath) {
  const parentPaths = getParentPaths(filePath);
  for (const parentPath of parentPaths) {
    const cache = dirPathTypeModuleCache.get(parentPath);
    if (cache !== void 0) {
      return cache;
    }
  }
  for (const parentPath of parentPaths) {
    try {
      const pkgPath = posixPath.join(parentPath, "package.json");
      const pkgJson = fs2.readFileSync(pkgPath, "utf8");
      const pkg = JSON.parse(pkgJson);
      const maybeModulePath = pkg.module ? posixPath.join(parentPath, pkg.module) : "";
      const cache = pkg.type === "module" || maybeModulePath === filePath;
      dirPathTypeModuleCache.set(parentPath, cache);
      return cache;
    } catch (e) {
      if (!isFileNotFoundError(e)) {
        throw e;
      }
    }
  }
  return false;
}
await cjsModuleLexer.init();
async function getCjsNamedExports(vite, filePath, contents, seen = /* @__PURE__ */ new Set()) {
  const { exports, reexports } = cjsModuleLexer.parse(contents);
  const result = new Set(exports);
  for (const reexport of reexports) {
    const resolved = await viteResolve(
      vite,
      reexport,
      filePath,
      /* isRequire */
      true
    );
    if (seen.has(resolved)) {
      continue;
    }
    try {
      const resolvedContents = fs2.readFileSync(resolved, "utf8");
      seen.add(filePath);
      const resolvedNames = await getCjsNamedExports(
        vite,
        resolved,
        resolvedContents,
        seen
      );
      seen.delete(filePath);
      for (const name of resolvedNames) {
        result.add(name);
      }
    } catch (e) {
      if (!isFileNotFoundError(e)) {
        throw e;
      }
    }
  }
  result.delete("default");
  result.delete("__esModule");
  return result;
}
function withSourceUrl(contents, url) {
  if (contents.lastIndexOf("//# sourceURL=") !== -1) {
    return contents;
  }
  const sourceURL = `
`;
  return contents + sourceURL;
}
function withImportMetaUrl(contents, url) {
  return contents.replaceAll("import.meta.url", JSON.stringify(url.toString()));
}
var jsExtensions = [".js", ".mjs", ".cjs"];
function maybeGetTargetFilePath(target) {
  if (isFile(target)) {
    return target;
  }
  for (const extension of jsExtensions) {
    const targetWithExtension = target + extension;
    if (fs2.existsSync(targetWithExtension)) {
      return targetWithExtension;
    }
  }
  if (target.endsWith(disableCjsEsmShimSuffix)) {
    return target;
  }
  if (isDirectory(target)) {
    return maybeGetTargetFilePath(target + "/index");
  }
}
function getApproximateSpecifier(target, referrerDir) {
  if (/^(node|cloudflare|workerd):/.test(target)) {
    return target;
  }
  return posixPath.relative(referrerDir, target);
}
async function viteResolve(vite, specifier, referrer, isRequire) {
  const resolved = await vite.pluginContainer.resolveId(specifier, referrer, {
    ssr: true,
    // https://github.com/vitejs/vite/blob/v5.1.4/packages/vite/src/node/plugins/resolve.ts#L178-L179
    custom: { "node-resolve": { isRequire } }
  });
  if (resolved === null) {
    if (isRequire && specifier[0] === ".") {
      return require2.resolve(specifier, { paths: [referrer] });
    }
    throw new Error("Not found");
  }
  if (resolved.id === "__vite-browser-external") {
    return emptyLibPath;
  }
  if (resolved.external) {
    let { id } = resolved;
    if (workerdBuiltinModules.has(id)) {
      return `/${id}`;
    }
    if (id.startsWith("node:")) {
      throw new Error("Not found");
    }
    id = `node:${id}`;
    if (workerdBuiltinModules.has(id)) {
      return `/${id}`;
    }
    return id;
  }
  return trimViteVersionHash(resolved.id);
}
async function resolve(vite, method, target, specifier, referrer) {
  const referrerDir = posixPath.dirname(referrer);
  let filePath = maybeGetTargetFilePath(target);
  if (filePath !== void 0) {
    return filePath;
  }
  if (referrerDir !== "/" && workerdBuiltinModules.has(specifier)) {
    return `/${specifier}`;
  }
  const specifierLibPath = posixPath.join(
    libPath,
    specifier.replaceAll(":", "/")
  );
  filePath = maybeGetTargetFilePath(specifierLibPath);
  if (filePath !== void 0) {
    return filePath;
  }
  return viteResolve(vite, specifier, referrer, method === "require");
}
function buildRedirectResponse(filePath) {
  if (isWindows && filePath[0] !== "/") {
    filePath = `/${filePath}`;
  }
  return new Response2(null, { status: 301, headers: { Location: filePath } });
}
function maybeGetForceTypeModuleContents(filePath) {
  const match = forceModuleTypeRegexp.exec(filePath);
  if (match === null) {
    return;
  }
  filePath = trimSuffix(match[0], filePath);
  const type = match[1];
  const contents = fs2.readFileSync(filePath);
  switch (type) {
    case "ESModule":
      return { esModule: contents.toString() };
    case "CommonJS":
      return { commonJsModule: contents.toString() };
    case "Text":
      return { text: contents.toString() };
    case "Data":
      return { data: contents };
    case "CompiledWasm":
      return { wasm: contents };
    case "PythonModule":
      return { pythonModule: contents.toString() };
    case "PythonRequirement":
      return { pythonRequirement: contents.toString() };
    default: {
      const exhaustive = type;
      assert3.fail(`Unreachable: ${exhaustive} modules are unsupported`);
    }
  }
}
function buildModuleResponse(target, contents) {
  let name = target;
  if (!isWindows) {
    name = posixPath.relative("/", target);
  }
  assert3(name[0] !== "/");
  const result = { name };
  for (const key in contents) {
    const value = contents[key];
    result[key] = value instanceof Uint8Array ? Array.from(value) : value;
  }
  return Response2.json(result);
}
async function load(vite, logBase, method, target, specifier, filePath) {
  if (target !== filePath) {
    if (method === "require" && !specifier.startsWith("node:")) {
      filePath += disableCjsEsmShimSuffix;
    }
    debuglog(logBase, "redirect:", filePath);
    return buildRedirectResponse(filePath);
  }
  if (filePath.endsWith(".wasm")) {
    filePath += `?mf_vitest_force=CompiledWasm`;
  }
  const maybeContents = maybeGetForceTypeModuleContents(filePath);
  if (maybeContents !== void 0) {
    debuglog(logBase, "forced:", filePath);
    return buildModuleResponse(target, maybeContents);
  }
  const disableCjsEsmShim = filePath.endsWith(disableCjsEsmShimSuffix);
  if (disableCjsEsmShim) {
    filePath = trimSuffix(disableCjsEsmShimSuffix, filePath);
  }
  const isEsm = filePath.endsWith(".mjs") || filePath.endsWith(".js") && isWithinTypeModuleContext(filePath);
  let contents = fs2.readFileSync(filePath, "utf8");
  const targetUrl = pathToFileURL(target);
  contents = withSourceUrl(contents, targetUrl);
  if (isEsm) {
    contents = withImportMetaUrl(contents, targetUrl);
    debuglog(logBase, "esm:", filePath);
    return buildModuleResponse(target, { esModule: contents });
  }
  const insertCjsEsmShim = method === "import" || specifier.startsWith("node:");
  if (insertCjsEsmShim && !disableCjsEsmShim) {
    const fileName = posixPath.basename(filePath);
    const disableShimSpecifier = `./${fileName}${disableCjsEsmShimSuffix}`;
    const quotedDisableShimSpecifier = JSON.stringify(disableShimSpecifier);
    let esModule = `import mod from ${quotedDisableShimSpecifier}; export default mod;`;
    for (const name of await getCjsNamedExports(vite, filePath, contents)) {
      esModule += ` export const ${name} = mod.${name};`;
    }
    debuglog(logBase, "cjs-esm-shim:", filePath);
    return buildModuleResponse(target, { esModule });
  }
  debuglog(logBase, "cjs:", filePath);
  return buildModuleResponse(target, { commonJsModule: contents });
}
async function handleModuleFallbackRequest(vite, request) {
  const method = request.headers.get("X-Resolve-Method");
  assert3(method === "import" || method === "require");
  const url = new URL(request.url);
  let target = url.searchParams.get("specifier");
  let referrer = url.searchParams.get("referrer");
  assert3(target !== null, "Expected specifier search param");
  assert3(referrer !== null, "Expected referrer search param");
  const referrerDir = posixPath.dirname(referrer);
  let specifier = getApproximateSpecifier(target, referrerDir);
  if (specifier.startsWith("file:")) {
    specifier = fileURLToPath(specifier);
  }
  if (isWindows) {
    if (target[0] === "/") {
      target = target.substring(1);
    }
    if (referrer[0] === "/") {
      referrer = referrer.substring(1);
    }
  }
  const quotedTarget = JSON.stringify(target);
  const logBase = `${method}(${quotedTarget}) relative to ${referrer}:`;
  try {
    const filePath = await resolve(vite, method, target, specifier, referrer);
    return await load(vite, logBase, method, target, specifier, filePath);
  } catch (e) {
    debuglog(logBase, "error:", e);
    console.error(
      `[vitest-pool-workers] Failed to ${method} ${JSON.stringify(target)} from ${JSON.stringify(referrer)}.`,
      "To resolve this, try bundling the relevant dependency with Vite.",
      "For more details, refer to https://developers.cloudflare.com/workers/testing/vitest-integration/known-issues/#module-resolution"
    );
  }
  return new Response2(null, { status: 404 });
}

// src/pool/index.ts
assert4(
  typeof __vite_ssr_import__ === "undefined",
  "Expected `@cloudflare/vitest-pool-workers` not to be transformed by Vite"
);
function structuredSerializableStringify(value) {
  if (value && typeof value === "object" && "r" in value && value.r && typeof value.r === "object" && "map" in value.r && value.r.map) {
    delete value.r.map;
  }
  return stringify(value, structuredSerializableReducers);
}
function structuredSerializableParse(value) {
  return parse(value, structuredSerializableRevivers);
}
var debuglog2 = util2.debuglog(
  "vitest-pool-workers:index",
  (fn) => debuglog2 = fn
);
var log2 = new Log2(LogLevel2.VERBOSE, { prefix: "vpw" });
var mfLog = new Log2(LogLevel2.WARN);
var __filename2 = fileURLToPath2(import.meta.url);
var __dirname2 = path5.dirname(__filename2);
var DIST_PATH = path5.resolve(__dirname2, "..");
var POOL_WORKER_PATH = path5.join(DIST_PATH, "worker/index.mjs");
var symbolizerWarning = "warning: Not symbolizing stack traces because $LLVM_SYMBOLIZER is not set.";
var ignoreMessages = [
  // Not user actionable
  // TODO(someday): this is normal operation and really shouldn't error
  "disconnected: operation canceled",
  "disconnected: worker_do_not_log; Request failed due to internal error",
  "disconnected: WebSocket was aborted"
];
function trimSymbolizerWarning(chunk) {
  return chunk.includes(symbolizerWarning) ? chunk.substring(chunk.indexOf("\n") + 1) : chunk;
}
function handleRuntimeStdio(stdout, stderr) {
  stdout.on("data", (chunk) => {
    process.stdout.write(chunk);
  });
  stderr.on("data", (chunk) => {
    const str = trimSymbolizerWarning(chunk.toString());
    if (ignoreMessages.some((message) => str.includes(message))) {
      return;
    }
    process.stderr.write(str);
  });
}
function forEachMiniflare(mfs, callback) {
  if (mfs instanceof Miniflare) {
    return callback(mfs);
  }
  const promises = [];
  for (const mf of mfs.values()) {
    promises.push(callback(mf));
  }
  return Promise.all(promises);
}
var allProjects = /* @__PURE__ */ new Map();
function getRunnerName(project, testFile) {
  const name = `${WORKER_NAME_PREFIX}runner-${project.getName().replace(/[^a-z0-9-]/gi, "_")}`;
  if (testFile === void 0) {
    return name;
  }
  const testFileHash = crypto.createHash("sha1").update(testFile).digest("hex");
  testFile = testFile.replace(/[^a-z0-9-]/gi, "_");
  return `${name}-${testFileHash}-${testFile}`;
}
function isDurableObjectDesignatorToSelf(value) {
  if (typeof value === "string") {
    return true;
  }
  return typeof value === "object" && value !== null && "className" in value && typeof value.className === "string" && (!("scriptName" in value) || value.scriptName === void 0);
}
function isWorkflowDesignatorToSelf(value, currentScriptName) {
  return typeof value === "object" && value !== null && "className" in value && typeof value.className === "string" && (!("scriptName" in value) || value.scriptName === void 0 || value.scriptName === currentScriptName);
}
function getDurableObjectDesignators(options) {
  const result = /* @__PURE__ */ new Map();
  const durableObjects = options.miniflare?.durableObjects ?? {};
  for (const [key, designator] of Object.entries(durableObjects)) {
    if (typeof designator === "string") {
      result.set(key, { className: designator });
    } else if (typeof designator.unsafeUniqueKey !== "symbol") {
      result.set(key, {
        className: designator.className,
        scriptName: designator.scriptName,
        unsafeUniqueKey: designator.unsafeUniqueKey
      });
    }
  }
  for (const unboundDurableObject of options.miniflare?.additionalUnboundDurableObjects ?? []) {
    if (typeof unboundDurableObject.unsafeUniqueKey !== "symbol") {
      result.set(unboundDurableObject.className, {
        className: unboundDurableObject.className,
        scriptName: unboundDurableObject.scriptName,
        unsafeUniqueKey: unboundDurableObject.unsafeUniqueKey
      });
    }
  }
  return result;
}
var POOL_WORKER_DIR = path5.dirname(POOL_WORKER_PATH);
var USER_OBJECT_MODULE_NAME = "__VITEST_POOL_WORKERS_USER_OBJECT";
var USER_OBJECT_MODULE_PATH = path5.join(
  POOL_WORKER_DIR,
  USER_OBJECT_MODULE_NAME
);
var DEFINES_MODULE_PATH = path5.join(
  POOL_WORKER_DIR,
  "__VITEST_POOL_WORKERS_DEFINES"
);
function getDurableObjectClasses(worker) {
  const result = /* @__PURE__ */ new Set();
  for (const designator of Object.values(worker.durableObjects ?? {})) {
    if (isDurableObjectDesignatorToSelf(designator)) {
      result.add(
        typeof designator === "string" ? designator : designator.className
      );
    }
  }
  for (const designator of worker.additionalUnboundDurableObjects ?? []) {
    result.add(designator.className);
  }
  return result;
}
function getWranglerWorkerName(relativeWranglerConfigPath) {
  if (!relativeWranglerConfigPath) {
    return void 0;
  }
  const wranglerConfigObject = experimental_readRawConfig({
    config: relativeWranglerConfigPath
  });
  return wranglerConfigObject.rawConfig.name;
}
function updateWorkflowsScriptNames(runnerWorker, wranglerWorkerName) {
  const workflows = runnerWorker.workflows;
  if (!workflows || wranglerWorkerName === void 0) {
    return;
  }
  for (const workflow of Object.values(workflows)) {
    if (workflow.scriptName === wranglerWorkerName) {
      delete workflow.scriptName;
    }
  }
}
function getWorkflowClasses(worker, relativeWranglerConfigPath) {
  const result = /* @__PURE__ */ new Set();
  if (worker.workflows === void 0) {
    return result;
  }
  for (const key of Object.keys(worker.workflows)) {
    const designator = worker.workflows[key];
    let workerName;
    const wranglerWorkerName = getWranglerWorkerName(
      relativeWranglerConfigPath
    );
    if (wranglerWorkerName && designator.scriptName === wranglerWorkerName) {
      workerName = wranglerWorkerName;
    } else {
      workerName = worker.name;
    }
    if (isWorkflowDesignatorToSelf(designator, workerName)) {
      result.add(designator.className);
      worker.workflows[key] = { ...designator };
    }
  }
  return result;
}
var SELF_NAME_BINDING = "__VITEST_POOL_WORKERS_SELF_NAME";
var SELF_SERVICE_BINDING = "__VITEST_POOL_WORKERS_SELF_SERVICE";
var LOOPBACK_SERVICE_BINDING = "__VITEST_POOL_WORKERS_LOOPBACK_SERVICE";
var RUNNER_OBJECT_BINDING = "__VITEST_POOL_WORKERS_RUNNER_OBJECT";
async function buildProjectWorkerOptions(project) {
  const relativeWranglerConfigPath = maybeApply(
    (v) => path5.relative("", v),
    project.options.wrangler?.configPath
  );
  const runnerWorker = project.options.miniflare ?? {};
  runnerWorker.name = getRunnerName(project.project);
  runnerWorker.bindings ??= {};
  runnerWorker.bindings[SELF_NAME_BINDING] = runnerWorker.name;
  runnerWorker.compatibilityFlags ??= [];
  if (runnerWorker.compatibilityDate === void 0) {
    runnerWorker.compatibilityDate ??= supportedCompatibilityDate;
    log2.info(
      `No compatibility date was provided for project ${project.relativePath}, defaulting to latest supported date ${runnerWorker.compatibilityDate}.`
    );
  }
  const flagAssertions = new CompatibilityFlagAssertions({
    compatibilityDate: runnerWorker.compatibilityDate,
    compatibilityFlags: runnerWorker.compatibilityFlags,
    optionsPath: `${OPTIONS_PATH}.miniflare`,
    relativeProjectPath: project.relativePath.toString(),
    relativeWranglerConfigPath
  });
  const assertions = [
    () => flagAssertions.assertIsEnabled({
      enableFlag: "export_commonjs_default",
      disableFlag: "export_commonjs_namespace",
      defaultOnDate: "2022-10-31"
    })
  ];
  for (const assertion of assertions) {
    const result = assertion();
    if (!result.isValid) {
      throw new Error(result.errorMessage);
    }
  }
  const { hasNoNodejsCompatV2Flag, mode } = getNodeCompat(
    runnerWorker.compatibilityDate,
    runnerWorker.compatibilityFlags
  );
  if (mode !== "v2") {
    if (hasNoNodejsCompatV2Flag) {
      runnerWorker.compatibilityFlags.splice(
        runnerWorker.compatibilityFlags.indexOf("no_nodejs_compat_v2"),
        1
      );
    }
    runnerWorker.compatibilityFlags.push("nodejs_compat_v2");
  }
  if (!runnerWorker.compatibilityFlags.includes("unsafe_module")) {
    runnerWorker.compatibilityFlags.push("unsafe_module");
  }
  ensureFeature(runnerWorker.compatibilityFlags, "nodejs_tty_module");
  ensureFeature(runnerWorker.compatibilityFlags, "nodejs_fs_module");
  ensureFeature(runnerWorker.compatibilityFlags, "nodejs_http_modules");
  ensureFeature(runnerWorker.compatibilityFlags, "nodejs_perf_hooks_module");
  runnerWorker.unsafeEvalBinding = "__VITEST_POOL_WORKERS_UNSAFE_EVAL";
  runnerWorker.unsafeUseModuleFallbackService = true;
  runnerWorker.serviceBindings ??= {};
  runnerWorker.serviceBindings[SELF_SERVICE_BINDING] = kCurrentWorker;
  runnerWorker.serviceBindings[LOOPBACK_SERVICE_BINDING] = handleLoopbackRequest;
  runnerWorker.durableObjects ??= {};
  const durableObjectClassNames = getDurableObjectClasses(runnerWorker);
  const workflowClassNames = getWorkflowClasses(
    runnerWorker,
    relativeWranglerConfigPath
  );
  const selfWorkerExports = [];
  if (flagAssertions.isEnabled(
    "enable_ctx_exports",
    "disable_ctx_exports",
    "2025-11-17"
  )) {
    try {
      const resolvedMain = maybeGetResolvedMainPath(project);
      const guessedExports = await guessWorkerExports(
        resolvedMain,
        project.options.additionalExports
      );
      for (const [exportName, exportType] of guessedExports) {
        switch (exportType) {
          case "DurableObject":
            durableObjectClassNames.add(exportName);
            break;
          case "WorkflowEntrypoint":
            workflowClassNames.add(exportName);
            break;
          case "WorkerEntrypoint":
          case null:
            selfWorkerExports.push(exportName);
        }
      }
    } catch (e) {
      const message = `Failed to statically analyze the exports of the main Worker entry-point "${project.options.main}"
More details: ${e}`;
      for (const line of message.split("\n")) {
        log2.warn(line);
      }
    }
  }
  const workerEntrypointExports = selfWorkerExports.filter(
    (name) => name !== "default" && name !== "__esModule" && !durableObjectClassNames.has(name) && !workflowClassNames.has(name)
  );
  const wrappers = [
    'import { createWorkerEntrypointWrapper, createDurableObjectWrapper, createWorkflowEntrypointWrapper } from "cloudflare:test-internal";'
  ];
  for (const entrypointName of workerEntrypointExports.sort()) {
    const quotedEntrypointName = JSON.stringify(entrypointName);
    const wrapper = `export const ${entrypointName} = createWorkerEntrypointWrapper(${quotedEntrypointName});`;
    wrappers.push(wrapper);
  }
  for (const className of Array.from(durableObjectClassNames).sort()) {
    const quotedClassName = JSON.stringify(className);
    const wrapper = `export const ${className} = createDurableObjectWrapper(${quotedClassName});`;
    wrappers.push(wrapper);
  }
  for (const className of Array.from(workflowClassNames).sort()) {
    const quotedClassName = JSON.stringify(className);
    const wrapper = `export const ${className} = createWorkflowEntrypointWrapper(${quotedClassName});`;
    wrappers.push(wrapper);
  }
  runnerWorker.durableObjects[RUNNER_OBJECT_BINDING] = {
    className: "__VITEST_POOL_WORKERS_RUNNER_DURABLE_OBJECT__",
    // Make the runner object ephemeral, so it doesn't write any `.sqlite` files
    // that would disrupt stacked storage because we prevent eviction
    unsafeUniqueKey: kUnsafeEphemeralUniqueKey,
    unsafePreventEviction: true
  };
  const defines = `export default {
		${Object.entries(project.options.defines ?? {}).map(([key, value]) => `${JSON.stringify(key)}: ${value}`).join(",\n")}
	};
	`;
  if ("script" in runnerWorker) {
    delete runnerWorker.script;
  }
  if ("scriptPath" in runnerWorker) {
    delete runnerWorker.scriptPath;
  }
  const modulesRoot = process.platform === "win32" ? "Z:\\" : "/";
  runnerWorker.modulesRoot = modulesRoot;
  runnerWorker.modules = [
    {
      type: "ESModule",
      path: path5.join(modulesRoot, POOL_WORKER_PATH),
      contents: fs3.readFileSync(POOL_WORKER_PATH)
    },
    {
      type: "ESModule",
      path: path5.join(modulesRoot, USER_OBJECT_MODULE_PATH),
      contents: wrappers.join("\n")
    },
    {
      type: "ESModule",
      path: path5.join(modulesRoot, DEFINES_MODULE_PATH),
      contents: defines
    },
    // The native workerd provided nodejs modules don't always support everything Vitest needs.
    // As a short-term fix, inject polyfills into the worker bundle that override the native modules.
    {
      type: "ESModule",
      path: path5.join(modulesRoot, "node:console"),
      contents: fs3.readFileSync(
        path5.join(DIST_PATH, `worker/node/console.mjs`)
      )
    },
    {
      type: "ESModule",
      path: path5.join(modulesRoot, "node:vm"),
      contents: fs3.readFileSync(path5.join(DIST_PATH, `worker/node/vm.mjs`))
    }
  ];
  const workers = [runnerWorker];
  if (runnerWorker.workers !== void 0) {
    for (let i = 0; i < runnerWorker.workers.length; i++) {
      const worker = runnerWorker.workers[i];
      if (typeof worker !== "object" || worker === null || !("name" in worker) || typeof worker.name !== "string" || worker.name === "") {
        throw new Error(
          `In project ${project.relativePath}, \`${OPTIONS_PATH}.miniflare.workers[${i}].name\` must be non-empty`
        );
      }
      if (worker.name.startsWith(WORKER_NAME_PREFIX)) {
        throw new Error(
          `In project ${project.relativePath}, \`${OPTIONS_PATH}.miniflare.workers[${i}].name\` must not start with "${WORKER_NAME_PREFIX}", got ${worker.name}`
        );
      }
      workers.push(worker);
    }
    delete runnerWorker.workers;
  }
  return workers;
}
var SHARED_MINIFLARE_OPTIONS = {
  log: mfLog,
  verbose: true,
  handleRuntimeStdio,
  unsafeStickyBlobs: true
};
var DEFAULT_INSPECTOR_PORT = 9229;
function getFirstAvailablePort(start) {
  return getPorts({ port: portNumbers(start, 65535) });
}
var moduleFallbackServices = /* @__PURE__ */ new WeakMap();
function getModuleFallbackService(ctx) {
  let service = moduleFallbackServices.get(ctx);
  if (service !== void 0) {
    return service;
  }
  service = handleModuleFallbackRequest.bind(void 0, ctx.vitenode.server);
  moduleFallbackServices.set(ctx, service);
  return service;
}
async function buildProjectMiniflareOptions(ctx, project) {
  const moduleFallbackService = getModuleFallbackService(ctx);
  const [runnerWorker, ...auxiliaryWorkers] = await buildProjectWorkerOptions(project);
  assert4(runnerWorker.name !== void 0);
  assert4(runnerWorker.name.startsWith(WORKER_NAME_PREFIX));
  let inspectorPort;
  if (ctx.config.inspector.enabled) {
    const userSpecifiedPort = ctx.config.inspector.port;
    if (userSpecifiedPort !== void 0) {
      const availablePort = await getFirstAvailablePort(userSpecifiedPort);
      if (availablePort !== userSpecifiedPort) {
        throw new Error(
          `Inspector port ${userSpecifiedPort} is not available. Either free up the port or remove the inspector port configuration to use an automatically assigned port.`
        );
      }
      inspectorPort = userSpecifiedPort;
    } else {
      inspectorPort = await getFirstAvailablePort(DEFAULT_INSPECTOR_PORT);
      if (inspectorPort !== DEFAULT_INSPECTOR_PORT) {
        log2.warn(
          `Default inspector port ${DEFAULT_INSPECTOR_PORT} not available, using ${inspectorPort} instead.`
        );
      }
    }
  }
  if (inspectorPort !== void 0 && !project.options.singleWorker) {
    log2.warn(`Tests run in singleWorker mode when the inspector is open.`);
    project.options.singleWorker = true;
  }
  if (project.options.singleWorker || project.options.isolatedStorage) {
    const wranglerWorkerName = getWranglerWorkerName(
      project.options.wrangler?.configPath
    );
    updateWorkflowsScriptNames(runnerWorker, wranglerWorkerName);
    return {
      ...SHARED_MINIFLARE_OPTIONS,
      inspectorPort,
      unsafeModuleFallbackService: moduleFallbackService,
      workers: [runnerWorker, ABORT_ALL_WORKER, ...auxiliaryWorkers]
    };
  } else {
    const testWorkers = [];
    for (const testFile of project.testFiles) {
      const testWorker = { ...runnerWorker };
      testWorker.name = getRunnerName(project.project, testFile);
      assert4(testWorker.bindings !== void 0);
      testWorker.bindings = { ...testWorker.bindings };
      testWorker.bindings[SELF_NAME_BINDING] = testWorker.name;
      const wranglerWorkerName = getWranglerWorkerName(
        project.options.wrangler?.configPath
      );
      updateWorkflowsScriptNames(testWorker, wranglerWorkerName);
      testWorkers.push(testWorker);
    }
    return {
      ...SHARED_MINIFLARE_OPTIONS,
      unsafeModuleFallbackService: moduleFallbackService,
      workers: [...testWorkers, ABORT_ALL_WORKER, ...auxiliaryWorkers]
    };
  }
}
async function getProjectMiniflare(ctx, project) {
  const mfOptions = await buildProjectMiniflareOptions(ctx, project);
  const changed = !util2.isDeepStrictEqual(project.previousMfOptions, mfOptions);
  project.previousMfOptions = mfOptions;
  const previousSingleInstance = project.mf instanceof Miniflare;
  const singleInstance = project.options.singleWorker || !project.options.isolatedStorage;
  if (project.mf !== void 0 && previousSingleInstance !== singleInstance) {
    log2.info(`Isolation changed for ${project.relativePath}, resetting...`);
    await forEachMiniflare(project.mf, (mf) => mf.dispose());
    project.mf = void 0;
  }
  if (project.mf === void 0) {
    if (singleInstance) {
      log2.info(
        `Starting single runtime for ${project.relativePath}${mfOptions.inspectorPort !== void 0 ? ` with inspector on port ${mfOptions.inspectorPort}` : ""}...`
      );
      project.mf = new Miniflare(mfOptions);
    } else {
      log2.info(`Starting isolated runtimes for ${project.relativePath}...`);
      project.mf = /* @__PURE__ */ new Map();
      for (const testFile of project.testFiles) {
        project.mf.set(testFile, new Miniflare(mfOptions));
      }
    }
    await forEachMiniflare(project.mf, (mf) => mf.ready);
  } else if (changed) {
    log2.info(`Options changed for ${project.relativePath}, updating...`);
    await forEachMiniflare(project.mf, (mf) => mf.setOptions(mfOptions));
  } else {
    log2.debug(`Reusing runtime for ${project.relativePath}...`);
  }
  return project.mf;
}
function maybeGetResolvedMainPath(project) {
  const projectPath = getProjectPath(project.project);
  const main = project.options.main;
  if (main === void 0) {
    return;
  }
  if (typeof projectPath === "string") {
    return path5.resolve(path5.dirname(projectPath), main);
  } else {
    return path5.resolve(main);
  }
}
async function runTests(ctx, mf, workerName, project, config, files, invalidates = [], method) {
  const workerPath = path5.join(ctx.distPath, "worker.js");
  const threadsWorkerPath = path5.join(ctx.distPath, "workers", "threads.js");
  ctx.state.clearFiles(project.project, files);
  const data = {
    pool: "threads",
    worker: pathToFileURL2(threadsWorkerPath).href,
    port: void 0,
    config,
    files,
    invalidates,
    environment: { name: "node", options: null },
    workerId: 0,
    projectName: project.project.getName(),
    providedContext: project.project.getProvidedContext()
  };
  const configPlugin = project.project.server.config.plugins.find(
    ({ name }) => name === "@cloudflare/vitest-pool-workers:config"
  );
  if (configPlugin !== void 0) {
    const api = configPlugin.api;
    api.setMain(project.options.main);
  }
  await waitForStorageReset(mf);
  const ns = await mf.getDurableObjectNamespace(
    RUNNER_OBJECT_BINDING,
    workerName
  );
  const stub = ns.get("singleton");
  const res = await stub.fetch("http://placeholder", {
    headers: {
      Upgrade: "websocket",
      "MF-Vitest-Worker-Data": structuredSerializableStringify({
        filePath: pathToFileURL2(workerPath).href,
        name: method,
        data,
        cwd: process.cwd()
      })
    }
  });
  const webSocket = res.webSocket;
  assert4(webSocket !== null);
  const chunkingSocket = createChunkingSocket({
    post(message) {
      webSocket.send(message);
    },
    on(listener) {
      webSocket.addEventListener("message", (event2) => {
        listener(event2.data);
      });
    }
  });
  const rules = project.options.miniflare?.modulesRules;
  const compiledRules = compileModuleRules(rules ?? []);
  const localRpcFunctions = createMethodsRPC(project.project, {
    cacheFs: false
  });
  const patchedLocalRpcFunctions = {
    ...localRpcFunctions,
    async fetch(...args) {
      const specifier = args[0];
      if (specifier !== "cloudflare:test" && (/^(cloudflare|workerd):/.test(specifier) || workerdBuiltinModules.has(specifier))) {
        return { externalize: specifier };
      }
      const maybeRule = compiledRules.find(
        (rule) => testRegExps(rule.include, specifier)
      );
      if (maybeRule !== void 0) {
        const externalize = specifier + `?mf_vitest_force=${maybeRule.type}`;
        return { externalize };
      }
      return localRpcFunctions.fetch(...args);
    }
  };
  let startupError;
  const rpc = createBirpc(patchedLocalRpcFunctions, {
    eventNames: ["onCancel"],
    post(value) {
      if (webSocket.readyState === WebSocket.READY_STATE_OPEN) {
        debuglog2("POOL-->WORKER", value);
        chunkingSocket.post(structuredSerializableStringify(value));
      } else {
        debuglog2("POOL--*      ", value);
      }
    },
    on(listener) {
      chunkingSocket.on((message) => {
        const value = structuredSerializableParse(message);
        debuglog2("POOL<--WORKER", value);
        if (typeof value === "object" && value !== null && "vitestPoolWorkersError" in value) {
          startupError = value.vitestPoolWorkersError;
        } else {
          listener(value);
        }
      });
    }
  });
  project.project.ctx.onCancel((reason) => rpc.onCancel(reason));
  webSocket.accept();
  const [event] = await events.once(webSocket, "close");
  if (webSocket.readyState === WebSocket.READY_STATE_CLOSING) {
    if (event.code === 1005) {
      webSocket.close();
    } else {
      webSocket.close(event.code, event.reason);
    }
  }
  if (event.code !== 1e3) {
    throw startupError ?? new Error("Failed to run tests");
  }
  debuglog2("DONE", files);
}
function getPackageJson(dirPath) {
  while (true) {
    const pkgJsonPath = path5.join(dirPath, "package.json");
    try {
      const contents = fs3.readFileSync(pkgJsonPath, "utf8");
      return JSON.parse(contents);
    } catch (e) {
      if (!isFileNotFoundError(e)) {
        throw e;
      }
    }
    const nextDirPath = path5.dirname(dirPath);
    if (nextDirPath === dirPath) {
      return;
    }
    dirPath = nextDirPath;
  }
}
function assertCompatibleVitestVersion(ctx) {
  const poolPkgJson = getPackageJson(__dirname2);
  const vitestPkgJson = getPackageJson(ctx.distPath);
  assert4(
    poolPkgJson !== void 0,
    "Expected to find `package.json` for `@cloudflare/vitest-pool-workers`"
  );
  assert4(
    vitestPkgJson !== void 0,
    "Expected to find `package.json` for `vitest`"
  );
  const expectedVitestVersion = poolPkgJson.peerDependencies?.vitest;
  const actualVitestVersion = vitestPkgJson.version;
  assert4(
    expectedVitestVersion !== void 0,
    "Expected to find `@cloudflare/vitest-pool-workers`'s `vitest` version constraint"
  );
  assert4(
    actualVitestVersion !== void 0,
    "Expected to find `vitest`'s version"
  );
  if (!(0, import_satisfies.default)(actualVitestVersion, expectedVitestVersion)) {
    const message = [
      `You're running \`vitest@${actualVitestVersion}\`, but this version of \`@cloudflare/vitest-pool-workers\` only officially supports \`vitest ${expectedVitestVersion}\`.`,
      "`@cloudflare/vitest-pool-workers` currently depends on internal Vitest APIs that are not protected by semantic-versioning guarantees.",
      `Your tests may work without issue, but we can not guarantee compatibility outside of the above version range.`
    ].join("\n");
    log2.warn(message);
  }
}
var warnedUnsupportedInspectorOptions = false;
function validateInspectorConfig(config) {
  if (config.inspector.host) {
    throw new TypeError(
      "Customizing inspector host is not supported with vitest-pool-workers."
    );
  }
  if (config.inspector.enabled && !warnedUnsupportedInspectorOptions) {
    if (config.inspectBrk) {
      log2.warn(
        `The "--inspect-brk" flag is not supported. Use "--inspect" instead.`
      );
    } else if (config.inspector.waitForDebugger) {
      log2.warn(
        `The "inspector.waitForDebugger" option is not supported. Insert a debugger statement if you need to pause execution.`
      );
    }
    warnedUnsupportedInspectorOptions = true;
  }
}
async function executeMethod(ctx, specs, invalidates, method) {
  const parsedProjectOptions = /* @__PURE__ */ new Set();
  for (const [project, testFile] of specs) {
    const projectName = project.getName();
    let workersProject = allProjects.get(projectName);
    if (workersProject === void 0) {
      workersProject = {
        project,
        options: await parseProjectOptions(project),
        testFiles: /* @__PURE__ */ new Set(),
        relativePath: getRelativeProjectPath(project)
      };
      allProjects.set(projectName, workersProject);
    } else if (!parsedProjectOptions.has(project)) {
      workersProject.project = project;
      workersProject.options = await parseProjectOptions(project);
      workersProject.relativePath = getRelativeProjectPath(project);
    }
    workersProject.testFiles.add(testFile);
    parsedProjectOptions.add(project);
  }
  const resultPromises = [];
  const filesByProject = /* @__PURE__ */ new Map();
  for (const [project, file] of specs) {
    let group = filesByProject.get(project);
    if (group === void 0) {
      filesByProject.set(project, group = []);
    }
    group.push(file);
  }
  for (const [workspaceProject, files] of filesByProject) {
    const project = allProjects.get(workspaceProject.getName());
    assert4(project !== void 0);
    const options = project.options;
    const config = workspaceProject.getSerializableConfig();
    config.runner = "cloudflare:test-runner";
    config.fakeTimers.toFake = config.fakeTimers.toFake?.filter(
      (timerMethod) => timerMethod !== "setImmediate" && timerMethod !== "clearImmediate"
    );
    validateInspectorConfig(config);
    config.inspector = {
      enabled: false
    };
    config.poolOptions = {
      // @ts-expect-error Vitest provides no way to extend this type
      threads: {
        // Allow workers to be re-used by removing the isolation requirement
        isolate: false
      },
      workers: {
        // Include resolved `main` if defined, and the names of Durable Object
        // bindings that point to classes in the current isolate in the
        // serialized config
        main: maybeGetResolvedMainPath(project),
        // Include designators of all Durable Object namespaces bound in the
        // runner worker. We'll use this to list IDs in a namespace. We'll
        // also use this to check Durable Object test runner helpers are
        // only used with classes defined in the current worker, as these
        // helpers rely on wrapping the object.
        durableObjectBindingDesignators: getDurableObjectDesignators(
          project.options
        ),
        // Include whether isolated storage has been enabled for this
        // project, so we know whether to call out to the loopback service
        // to push/pop the storage stack between tests.
        isolatedStorage: project.options.isolatedStorage
      }
    };
    const mf = await getProjectMiniflare(ctx, project);
    if (options.singleWorker) {
      assert4(mf instanceof Miniflare, "Expected single instance");
      const name = getRunnerName(workspaceProject);
      resultPromises.push(
        runTests(ctx, mf, name, project, config, files, invalidates, method)
      );
    } else if (options.isolatedStorage) {
      assert4(mf instanceof Map, "Expected multiple isolated instances");
      const name = getRunnerName(workspaceProject);
      for (const file of files) {
        const fileMf = mf.get(file);
        assert4(fileMf !== void 0);
        resultPromises.push(
          runTests(
            ctx,
            fileMf,
            name,
            project,
            config,
            [file],
            invalidates,
            method
          )
        );
      }
    } else {
      assert4(mf instanceof Miniflare, "Expected single instance");
      for (const file of files) {
        const name = getRunnerName(workspaceProject, file);
        resultPromises.push(
          runTests(ctx, mf, name, project, config, [file], invalidates, method)
        );
      }
    }
  }
  const results = await Promise.allSettled(resultPromises);
  const errors = results.filter((r) => r.status === "rejected").map((r) => r.reason);
  for (const project of allProjects.values()) {
    if (project.mf !== void 0) {
      void forEachMiniflare(project.mf, async (mf) => scheduleStorageReset(mf));
    }
  }
  if (errors.length > 0) {
    throw new AggregateError(
      errors,
      "Errors occurred while running tests. For more information, see serialized error."
    );
  }
}
function pool_default(ctx) {
  assertCompatibleVitestVersion(ctx);
  return {
    name: "vitest-pool-workers",
    async runTests(specs, invalidates) {
      await executeMethod(ctx, specs, invalidates, "run");
    },
    async collectTests(specs, invalidates) {
      await executeMethod(ctx, specs, invalidates, "collect");
    },
    async close() {
      log2.debug("Shutting down runtimes...");
      const promises = [];
      for (const project of allProjects.values()) {
        if (project.mf !== void 0) {
          promises.push(
            forEachMiniflare(project.mf, async (mf) => {
              await waitForStorageReset(mf);
              await mf.dispose();
            })
          );
        }
      }
      allProjects.clear();
      await Promise.all(promises);
    }
  };
}
function ensureFeature(compatibilityFlags, feature) {
  const flagToEnable = `enable_${feature}`;
  const flagToDisable = `disable_${feature}`;
  if (!compatibilityFlags.includes(flagToEnable)) {
    log2.debug(
      `Adding \`${flagToEnable}\` compatibility flag during tests as this feature is needed to support the Vitest runner.`
    );
    compatibilityFlags.push(flagToEnable);
  }
  if (compatibilityFlags.includes(flagToDisable)) {
    log2.info(
      `Removing \`${flagToDisable}\` compatibility flag during tests as that feature is needed to support the Vitest runner.`
    );
    compatibilityFlags.splice(compatibilityFlags.indexOf(flagToDisable), 1);
  }
}
export {
  pool_default as default
};
